---
title: "FigureYa126CorrelationHeatmap"
author: "小丫画图出品"
date: "2019-8-4"
output: html_document
---
欢迎关注“小丫画图”公众号，同名知识星球等你加入

小丫微信: epigenomics  E-mail: figureya@126.com

作者：大鱼海棠

单位：Research Center of Biostatistics and Computational Pharmacy, China Pharmaceutical University

小丫编辑校验

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 需求描述

根据相关系数排列的无聚类热图，展示和TIM-3基因表达（某一连续变量）相关的免疫基因（其他连续变量）在样本中的分布情况。使用TCGA GBM数据以及例文给出的基因集，复现原图。

![](example.png)

出自<https://www.tandfonline.com/doi/full/10.1080/2162402X.2017.1328339>

Figure 3. (C, D) Most immune response related genes were significantly positively correlated with TIM-3 expression, while there were still a small number of genes were significantly negatively correlated with Tim-3 expression.

## 应用场景

展示跟某一个基因呈正/负相关的多个基因的表达模式。

例如像例文这种，每列一个样本，每行一个基因。上半部分画出与目标基因的表达模式呈正相关的免疫相关基因，下半部分是负相关的。顶部展示样本的亚组、基因型或其他临床信息。

## 环境设置

使用国内镜像安装包

```{r}
options("repos"= c(CRAN="https://mirrors.tuna.tsinghua.edu.cn/CRAN/"))
options(BioC_mirror="http://mirrors.ustc.edu.cn/bioc/")
```

加载包

```{r}
library(clusterProfiler)
library(org.Hs.eg.db)
library(biomaRt)
library(pheatmap)
library(RColorBrewer)
Sys.setenv(LANGUAGE = "en") #显示英文报错信息
options(stringsAsFactors = FALSE) #禁止chr转成factor
```

## 输入文件

easy_input_expr.csv，TCGA-GBM.htseq_fpkm，基因表达谱。从XENA下载得到，值为log2（FPKM+1），为方便传输，小数点后保留2位。需要把基因名换成gene symbol。

gbm_tcga_pub2013_clinical_data.tsv，临床信息。从cBioPortal上下载得到：http://www.cbioportal.org/study/clinicalData?id=gbm_tcga_pub2013

easy_input_genes.txt，immune-response-related基因集。来自例文补充材料：Supplementary table1.xlsx

```{r}
### 读取GMB表达谱 ###
expr <- read.csv("easy_input_expr.csv", header = T, row.names = 1, stringsAsFactors = F, check.names = F)

rownames(expr) <- sapply(strsplit(rownames(expr),".",fixed = T),"[",1) # 基因ID取.号前的字符
colnames(expr) <- substr(colnames(expr),1,15)

# 利用bitr将entrez id 转为 gene symbol
ensembl2gs <- bitr(rownames(expr), fromType = "ENSEMBL", toType = "SYMBOL", OrgDb = org.Hs.eg.db)
ensembl2gs <- ensembl2gs[!duplicated(ensembl2gs$SYMBOL),] #去重

# 根据TCGA样品名的命名规律，提取出肿瘤样本
tum.sam <- colnames(expr)[which(substr(colnames(expr),14,15) == "01")]

# 生成新的表达谱
expr2 <- expr[ensembl2gs$ENSEMBL,tum.sam]
rownames(expr2) <- ensembl2gs$SYMBOL
expr2[1:3, 1:3]

### 读取临床数据并提取带临床信息的样本的表达谱 ###
Sinfo <- read.table("gbm_tcga_pub2013_clinical_data.tsv",sep = "\t",check.names = F,stringsAsFactors = F,header = T,row.names = NULL)
rownames(Sinfo) <- Sinfo$`Sample ID`
Sinfo <- Sinfo[intersect(colnames(expr2),rownames(Sinfo)),]
head(Sinfo)
# 提取带临床信息的样本对应的表达谱
expr2 <- expr2[,rownames(Sinfo)]

# 根据TIM-3的表达量，给样本排序（TIM-3即为HAVCR2）
sam.order <- sort(expr2["HAVCR2",],decreasing = F)

### 读取相关基因的表达量，用于画热图，例文画的是免疫相关基因 ###
genelist <- read.table("easy_input_genes.txt",sep = "\t",check.names = F,stringsAsFactors = F,header = T,row.names = NULL)
head(genelist)
imm.expr <- expr2[intersect(rownames(expr2), genelist$`GENE LIST`), names(sam.order)]
imm.expr[1:3, 1:3]
```

## 与基因集的相关性分析

```{r}
# 初始化相关性结果矩阵
cor.res <- NULL
for (i in 1:nrow(imm.expr)) {
  tmp <- cor.test(as.numeric(imm.expr[i,]),as.numeric(sam.order),method = "spearman") # 使用非参相关性分析
  cor.res <- rbind.data.frame(cor.res,data.frame(gene = rownames(imm.expr)[i], rho = tmp$estimate, p = tmp$p.value, stringsAsFactors = F))
}

pos.cor.gene <- cor.res[which(cor.res$rho > 0.3 & cor.res$p < 0.05), c("gene","rho")] # 显著正相关的基因
neg.cor.gene <- cor.res[which(cor.res$rho < -0.3 & cor.res$p < 0.05), c("gene","rho")] # 显著负相关性的基因（若用例文的-0.4cutoff，这里只能找到2个基因，最终图的效果可能不明显）

# 升序排列，这会让热图呈现出例文那样的楔形渐变的效果
pos.cor.gene <- pos.cor.gene[order(pos.cor.gene$rho,decreasing = F),] 
neg.cor.gene <- neg.cor.gene[order(neg.cor.gene$rho,decreasing = F),]
```

## 开始画图

```{r}
### 设置颜色 ###
heatmap.BlWtRd <- c("#6699CC","white","#FF3C38")
red    <- "#EA6767"
green  <- "#70C17A"
blue   <- "#445EAD"
grey   <- "#7A7A7A"
orange <- "#F7C07C"
yellow <- "#CEDC7C"

# 生成注释表
annCol <- data.frame("TIM-3" = as.numeric(sam.order[rownames(Sinfo)]),
                     "Subtype" = Sinfo$`Gene Expression Subtype`,
                     "IDH" = Sinfo$`IDH1 Mutation`,
                     row.names = rownames(Sinfo),
                     stringsAsFactors = F,
                     check.names = F) # 防止把“-”转换为“.”
annCol[is.na(annCol)] <- "NA"

# 生成注释颜色
annCol$IDH <- ifelse(annCol$IDH == "WT","WT",ifelse(annCol$IDH == "NA","NA","Mutant"))
annColors <- list()
annColors[["TIM-3"]] <- colorRampPalette(heatmap.BlWtRd)(128)
annColors[["IDH"]] <- c("WT"=red,"Mutant"=green,"NA"=grey)
annColors[["Subtype"]] <- c("Classical"=yellow,
                            "Mesenchymal"=red,
                            "Neural"=orange,
                            "Proneural"=blue,
                            "G-CIMP"=green,
                            "NA"=grey)

# 标准化并截断数据
indata <- t(scale(t(imm.expr[c(pos.cor.gene$gene,neg.cor.gene$gene),names(sam.order)]))) # 注意数据要根据TIM-3的表达顺序排列
indata[indata > 3]  <- 3
indata[indata < -3] <- -3

# 绘制热图
pheatmap(mat = indata,
         scale = "none", # 不标准化
         border_color = NA,
         color = colorRampPalette(heatmap.BlWtRd)(128), # 例文配色
         cluster_cols = F, # 列不聚类
         cluster_rows = F, # 行不聚类
         show_rownames = F, # 不显示行名
         show_colnames = F, # 不显示列名
         annotation_col = annCol[names(sam.order),], # 列注释（注意要根据TIM-3的表达顺序排列）
         annotation_colors = annColors, # 注释颜色
         gaps_row = length(pos.cor.gene$gene), # 在正相关基因结束后分割热图
         filename = "correlationHeatmap.pdf") 
```

![](correlationHeatmap.pdf)

```{r}
sessionInfo()
```