---
title: "FigureYa215DNAage"
author: "小丫画图出品"
date: "2021-1-15"
output: html_document
---
欢迎关注“小丫画图”公众号，回复“小白”，看小视频，实现点鼠标跑代码。

小丫微信: epigenomics  E-mail: figureya@126.com

作者：大鱼海棠

单位：中国药科大学国家天然药物重点实验室，生物统计与计算药学研究中心

小丫编辑校验

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述

求文中甲基化年龄的确定方法。

![](method.png)

出自<https://clinicalepigeneticsjournal.biomedcentral.com/articles/10.1186/s13148-020-0822-y>

# 应用场景

根据DNA甲基化beta值计算甲基化年龄，并与真实年龄（chronological age）做差值，计算甲基化年龄加速（Epigenetic age acceleration）。

> 注意：目前甲基化年龄计算只适用于450k和27k数据，850k不适用

接下来，例文把sample分为两组：DNAmAge-ACC（加速）和DNAmAge-DEC（减速），分别分析DNA甲基化年龄跟其他各项特征之间的关系：例如tumor vs normal，HPV16/18 expression，临床信息，immunoactivation，clinical outcomes。

使用本代码或分析思路请引用：

Lu X, Zhou Y, Meng J, et al. Epigenetic age acceleration of cervical squamous cell carcinoma converged to human papillomavirus 16/18 expression, immunoactivation, and favourable prognosis[J]. Clinical Epigenetics, 2020, 12(1): 1-13.

另外，例文中的Fig.1a可参考FigureYa163twoVarCor；
Fig.1e的画法可参考FigureYa162boxViolin；
Fig.2b的画法可参考FigureYa60GSEA_clusterProfiler；
Fig.4f可参考FigureYa99smoothHR。

大鱼海棠的更多作品看这里：<https://k.koudai.com/OFad8N0w>

# 环境设置

使用国内镜像安装包

```{r eval=FALSE}
options("repos"= c(CRAN="https://mirrors.tuna.tsinghua.edu.cn/CRAN/"))
options(BioC_mirror="http://mirrors.tuna.tsinghua.edu.cn/bioconductor/")
BiocManager::install("ChAMP")
```

加载包

```{r}
library(ChAMP)
library(wateRmelon)
library(data.table)

Sys.setenv(LANGUAGE = "en") #显示英文报错信息
options(stringsAsFactors = FALSE) #禁止chr转成factor
```

# 输入文件

easy_input_age.txt，临床真实年龄。

HumanMethylation450，CESC的DNA甲基化数据beta value。下载自<https://xenabrowser.net/datapages/?dataset=TCGA.CESC.sampleMap%2FHumanMethylation450&host=https%3A%2F%2Ftcga.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443>，已上传至微云<https://share.weiyun.com/o1g5N29h>

```{r}
## 读取临床真实年龄
age <- read.table("easy_input_age.txt",sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
head(age)

## 快速加载DNA甲基化数据
orgmeth <- fread("HumanMethylation450") 
orgmeth <- as.data.frame(orgmeth); rownames(orgmeth) <- orgmeth[,1]; orgmeth <- orgmeth[,-1]
orgmeth[1:3,1:3] # 查看一下数据
colnames(orgmeth) <- paste0("CESC",substr(colnames(orgmeth),8,15)) # 替换样本名和年龄匹配（这一步不是必须的）

tum.sam <- colnames(orgmeth[,which(substr(colnames(orgmeth),11,12) == "01")]) # 提取原位癌样本
nor.sam <- colnames(orgmeth[,which(substr(colnames(orgmeth),11,12) == "11")]) # 提取癌旁正常样本
orgmeth <- orgmeth[,c(tum.sam,nor.sam)] # 对数据做肿瘤、癌旁的排序
orgmeth <- as.data.frame(na.omit(orgmeth)) # 去除空值
```

# 计算甲基化年龄

## 甲基化数据过滤及标准化
```{r}
pd <- data.frame(Sample_Name = colnames(orgmeth),
                 Sample_Group = rep(c("T","N"),c(length(tum.sam),length(nor.sam))),
                 row.names = colnames(orgmeth),
                 stringsAsFactors = F)

myFilter <- champ.filter(beta = as.matrix(orgmeth),
                         arraytype = "450K", # 平台使用的是甲基化450K
                         pd = pd,
                         filterNoCG = T, # 非CG位点被移除
                         filterSNPs = T, # 靠近SNP位点的CpG被移除
                         filterMultiHit = T, # 探针对应多个位点的被移除
                         filterXY = T, # 移除性染色体
                         fixOutlier = T, # 修正极端值
                         autoimpute = F) # 不填补缺失值

# 速度较慢请耐心，我跑了30min左右
myFilter <- champ.norm(as.matrix(myFilter$beta), 
                       method = "BMIQ", # 使用BMIQ法对数据进行标准化
                       plotBMIQ = F, # 不绘制相关标准化图像
                       cores = 1, # 单核运行
                       arraytype = "450K") # 平台采用甲基化450K
save(myFilter,file = "meth.filter.norm.rda") # 保存数据，方便后续使用
```

## 计算甲基化年龄

```{r}
dnamage <- agep(myFilter[,tum.sam],
                coeff=NULL)
dnamage <- as.data.frame(dnamage); colnames(dnamage) <- "dnamage"

# 计算生理年龄和实际临床年龄的差异
dnamage$cage <- age[tum.sam,"Age"] # 加上临床年龄（clinical age）
dnamage$diff <- dnamage$dnamage - dnamage$cage 
# 定义DNAmAge-ACC（加速）和DNAmAge-DEC（减速）
# 若差异为正，则为甲基化年龄加速，否则为减速
dnamage$agestatus <- ifelse(dnamage$diff > 0, "Accelerated","Decelerated") 
dnamage <- as.data.frame(na.omit(dnamage))
sam.order <- rownames(dnamage)[order(dnamage$diff,decreasing = T)]
acc.num <- table(dnamage$agestatus)[1]
dec.num <- table(dnamage$agestatus)[2]
# 保存到文件
write.table(dnamage, "DNA methylation age and clinical age.txt",sep = "\t",row.names = T,col.names = NA,quote = F)
```

# 开始画图

用以上结果可以画出例文Fig. 3的中间部分，对比DNA甲基化年龄和真实年龄的acceleration

![](example.png)

Fig. 3 Differential DNA methylation pattern in CpG islands between two DNAm age groups. A total of 142 stringent hypermethylated probes and 7 hypomethylated probes were identified for the DNAmAge-ACC group. The heatmap based on DNA methylation M values demonstrates a co-occurrence of epigenetic age acceleration and immunoactivation as well as CpG island hypermethylation. DMPs: differentially methylated probes

```{r}
# 设置颜色
jco <- c("#2874C5","#EABF00","#868686","#C6524A","#80A7DE")
sun  <- "#E53435"
darkblue  <- "#1d00ff"

pdf("barplot for DNA methylation acceleration.pdf",width = 6,height = 4)
par(mai = c(1,0.5,0.3,0.1))
# 画x轴上方的甲基化年龄
barplot(dnamage[sam.order,"dnamage"],
        axes = F,
        col = rep(ggplot2::alpha(jco[2:1],0.8),c(acc.num,dec.num)),
        border = NA,
        space = 0,
        ylim = c(-80,150)) # 根据自己的数据酌情修改
# 画x轴下方的临床实际年龄
barplot(-dnamage[sam.order,"cage"],
        axes = F,
        col = rep(ggplot2::alpha(jco[2:1],0.8),c(acc.num,dec.num)),
        border = NA,
        space = 0,
        add = T)
# 补齐y轴
axis(side = 2,at = c(-80,0,150),labels =  c(-80,0,150))
# 添加甲基化与临床真实年龄的差异
barplot(dnamage[sam.order,"diff"],axes = F,col = rep(ggplot2::alpha(c(sun,darkblue)),c(acc.num,dec.num)),border = NA,space = 0,add = T)
# 添加一条拟合黑线
par(new=T)
lines(1:nrow(dnamage),dnamage[sam.order,"diff"],lwd=1.5)
# 添加文字
text(nrow(dnamage)/2, 140,
     "DNA methylation age",
     cex = 1.6)
text(nrow(dnamage)/2, -70,
     "Chronological age",
     cex = 1.6)
invisible(dev.off())
```

![](barplot for DNA methylation acceleration.pdf)

# Session Info

```{r}
sessionInfo()
```