---
title: "FigureYa208FPI"
author: "小丫画图出品"
date: "2020-11-27"
output: html_document
---
欢迎关注“小丫画图”公众号，回复“小白”，看小视频，实现点鼠标跑代码。

小丫微信: epigenomics  E-mail: figureya@126.com

作者：大鱼海棠

单位：中国药科大学国家天然药物重点实验室，生物统计与计算药学研究中心

小丫编辑校验

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述

想众筹铁死亡指数（FPI）的计算方法，根据基因表达，算出ssGSEA富集分数，再计算FPI，并用GSE121689数据集验证。

![](example.png)

Figure 2. The Relations between FPI and Histological Types and Molecular Subtypes among Cancers
(A) The different FPIs between tumor and normal tissues among cancers.

![](method1.png)

![](method2.png)

出自<https://linkinghub.elsevier.com/retrieve/pii/S2589004220304892>

# 应用场景

例文通过查文献收集基因，用GSVA计算正负两个基因集的富集分数，然后计算FPI。

可批量计算多组（TCGA pancancer RNA-seq）、也可以单独计算一个数据集（GEO表达谱数据）。

另外，不仅可以计算铁死亡指数FPI，还可以按照这个方法，替换为自己收集的某一功能相关的基因集，自己自定义一个某某指数～～～

# 环境设置

使用国内镜像安装包

```{r}
options("repos"= c(CRAN="https://mirrors.tuna.tsinghua.edu.cn/CRAN/"))
options(BioC_mirror="http://mirrors.tuna.tsinghua.edu.cn/bioconductor/")

```

加载包

```{r}
library(data.table)
library(GSVA)
library(ggplot2)
Sys.setenv(LANGUAGE = "en") #显示英文报错信息
options(stringsAsFactors = FALSE) #禁止chr转成factor
```

# 输入文件

大文件已上传到微云：<https://share.weiyun.com/c8GQyxR4>

## 下载TCGA pan-cancer数据

GSVA的输入数据可以是microarray data，也可以是RNA-seq count、CPM、RPKM、TPM。

原文：We calculate now GSVA enrichment scores for these gene sets using first the microarray data and then the **RNA-seq integer count** data. Note that the only requirement to do the latter is to set the argument **kcdf="Poisson"** which is "Gaussian" by default. Note, however, that if our RNA-seq derived expression levels would be **continous, such as log-CPMs, log-RPKMs or log-TPMs**, the the default value of the **kcdf argument should remain unchanged**. 出自<https://www.bioconductor.org/packages/release/bioc/vignettes/GSVA/inst/doc/GSVA.pdf>

- 表达矩阵和Gene mapping：
  - 这里跟FigureYa55panCancer_violin保持一致，从[XENA](https://xenabrowser.net/datapages/)下载UCSC Toil RNA-seq Recompute TPM：[TCGA Pan-Cancer (PANCAN) (41 datasets)
](https://xenabrowser.net/datapages/?cohort=TCGA%20Pan-Cancer%20(PANCAN)&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443)里的[TOIL RSEM tpm (n=10,535) UCSC Toil RNA-seq Recompute](https://toil.xenahubs.net/download/tcga_rsem_isoform_tpm.gz)，[Gene Mapping](https://toil.xenahubs.net/download/probeMap/gencode.v23.annotation.transcript.probemap)
  - 或者下载GDC pipeline的FPKM-UQ：[GDC Pan-Cancer (PANCAN) (17 datasets)](https://xenabrowser.net/datapages/?cohort=GDC%20Pan-Cancer%20(PANCAN)&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443) 里的[HTSeq - FPKM-UQ (n=11,768) GDC Hub](https://gdc.xenahubs.net/download/GDC-PANCAN.htseq_fpkm-uq.tsv.gz)，[Gene Mapping](https://gdc.xenahubs.net/download/gencode.v22.annotation.gene.probeMap)

- TCGA_phenotype_denseDataOnlyDownload.tsv，样本注释。[TCGA Pan-Cancer (PANCAN) (41 datasets) phenotype](https://xenabrowser.net/datapages/?cohort=TCGA%20Pan-Cancer%20(PANCAN)&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443)里的[sample type and primary disease (n=12,804) Pan-Cancer Atlas Hub](https://pancanatlas.xenahubs.net/download/TCGA_phenotype_denseDataOnlyDownload.tsv.gz)，只有癌症全称，没有缩写。

- samplepair.txt，癌症缩写跟全称的对应关系

```{r}
# 读取泛癌表达谱和基因名文件
# 文件较大，占内存，请谨慎操作
panexpr <- fread("tcga_RSEM_gene_tpm",sep = "\t",header = T,check.names = F,stringsAsFactors = F)
annopb <- read.delim("gencode.v23.annotation.gene.probemap",row.names = 1,check.names = F,stringsAsFactors = F,header = T,sep = "\t")

# 匹配基因名
panexpr <- as.data.frame(panexpr)
rownames(panexpr) <- panexpr$sample; panexpr <- panexpr[,-1]
comgene <- intersect(rownames(annopb), rownames(panexpr))
panexpr <- panexpr[comgene,]; annopb <- annopb[comgene,]
panexpr$genename <- annopb$gene; panexpr <- panexpr[!duplicated(panexpr$genename),]
rownames(panexpr) <- panexpr$genename; panexpr <- panexpr[,-ncol(panexpr)]
panexpr[1:3,1:3]

# 读取样本注释
tcgacase <- read.delim(file="TCGA_phenotype_denseDataOnlyDownload.tsv",header = T,row.names = NULL,check.names = F,stringsAsFactors = F,sep = "\t")
tcgacase[1:2,]

# 读取疾病缩写和全称
samplepair <- read.delim("samplepair.txt",as.is = T)
tissueinfo <- samplepair[,1:2]
tissueinfo[1:2,]
# 把疾病缩写加到样本注释里
colnames(tcgacase)[4] <- "Detail"
tcgacase <- merge(tcgacase, tissueinfo, by = "Detail", all.x = T)
rownames(tcgacase) <- tcgacase$sample

# 取出原位癌以及癌旁样本
tcgacase <- tcgacase[which(tcgacase$sample_type_id %in% c(1,11)),]

# 取出例文图中的癌症sample
tumors <- c("KIRC","THCA","COAD","STAD","CHOL","LUAD","HNSC","LIHC","KIRP","READ","PRAD","LUSC","GBM","ESCA","BLCA","PAAD","UCEC","BRCA","KICH","CESC")
tcgacase <- tcgacase[which(tcgacase$TCGA %in% tumors),]

# 提取既有表达数据又有样本注释信息的sample
comsam <- intersect(colnames(panexpr),rownames(tcgacase)) 
tcgacase <- tcgacase[comsam,] # TCGA的sample

# 提取pancancer表达矩阵
panexpr <- panexpr[,comsam]
# 输出到文件，可用于后续更多分析
#write.csv(panexpr, "output_pancancer_expr.csv", quote = F)
```

# 计算FPI

套用例文计算FPI的方法，你也可以自己定义个某某index/score。

1) 查文献定义了24个ferroptosis regulator genes (FRGs)

原文：In this study, the twenty four genes that were identified to play critical roles in regulating ferroptosis by pre- vious studies were defined as ferroptosis regulator genes (FRGs), 

2) 定义了两类基因：positive components (pos.comp)和negative components (neg.comp)。

原文：The index to represent the ferroptosis level was establish based on the expression data for genes of ferroptosis core machine including positive components of LPCAT3, ACSL4, NCOA4, ALOX15, GPX4, SLC3A2, SLC7A11, NFE2L2, NOX1, NOX3, NOX4, NOX5 and negative components of FDFT1, HMGCR, COQ10A, COQ10B. 

3) 用GSVA里的ssGSEA计算positively (pos.comp) or negatively (neg.comp) regulated ferroptosis的enrichment score，二者相减就是FPI。

原文：The enrichment score (ES) of gene set that positively or negatively regulated ferroptosis was calculated using single sample gene set enrichment analysis (ssGSEA) in the R package ‘GSVA’ (Hanzelmann et al., 2013), and the normalized differences between the ES of the positive components minus negative components was defined as the ferroptosis potential index (FPI) to computationally dissect the ferroptosis levels/trends of the tissue samples.

```{r}
# 用ssgsea计算
pos.comp <- c("LPCAT3","ACSL4","NCOA4","ALOX15","GPX4","SLC3A2","SLC7A11","NFE2L2","NOX1","NOX3","NOX4","NOX5")
neg.comp <- c("FDFT1","HMGCR","COQ10A","COQ10B")
fpi.sig <- list(pos.comp = pos.comp,
                neg.comp = neg.comp)

#写个循环批量计算出所有癌症的enrichment score
ssgsea.list <- list() # 初始化结果列表
for (t in tumors) {
  message("--analysis of ",t," starts...")
  tumsam <- rownames(tcgacase[which(tcgacase$TCGA == t & tcgacase$sample_type == "Primary Tumor"),])
  norsam <- rownames(tcgacase[which(tcgacase$TCGA == t & tcgacase$sample_type == "Solid Tissue Normal"),])
  
  # 每次计算一个癌症类型的enrichment score
  # 提取当前癌症的表达矩阵
  expr <- panexpr[,c(tumsam,norsam)]

  # 重新对表达谱对数化
  ## 以下步骤请根据自己表达谱的情况来
  expr <- 2^expr - 0.001 # 原始数据为log2(x+0.001)
  expr[expr < 0] <- 0 # 小于0的值拉到0
  expr <- log2(expr + 1) # 重新对数化
  expr <- expr[rowSums(expr) > 0,] # 去除全0值
  expr <- expr[apply(expr, 1, sd) > 0,] # 去除平坦值
  # expr <- t(scale(t(expr)))
  
  # 把处理好的每种癌症的表达矩阵保存到文件，可用于后续更多分析
  #write.csv(expr, paste0("normalized_expr_", t, ".csv"), quote = F)
  
  es <- gsva(expr = as.matrix(expr),
            gset.idx.list = fpi.sig,
            method = "ssgsea",
            ssgsea.norm = TRUE)
 
  ssgsea.list[[t]] <- es
  gc() # 释放内存
}

# 查看ssgsea.list里的结构
ssgsea.list$KIRC[,1:4]
# 可以看出，每种癌症有两行：pos.comp和neg.comp，接下来，用这两行相减，获得每个sample的PFI

# 计算FPI
fpi.res <- NULL
dirct <- c()
for (t in tumors) {
  tumsam <- rownames(tcgacase[which(tcgacase$TCGA == t & tcgacase$sample_type == "Primary Tumor"),])
  norsam <- rownames(tcgacase[which(tcgacase$TCGA == t & tcgacase$sample_type == "Solid Tissue Normal"),])
  
  es <- ssgsea.list[[t]]
  fpi <- scale(as.numeric(es[1,] - es[2,]))[,1] # 计算FPI并标准化
  names(fpi) <- c(tumsam, norsam)
  med.fpi.tum <- median(fpi[tumsam]) # 计算中位数
  med.fpi.nor <- median(fpi[norsam]) # 计算中位数
  wt <- wilcox.test(fpi[tumsam],fpi[norsam]) # 秩和检验
  
  # 若肿瘤FPI大于癌旁，则标签为黄色
  dirct <- c(dirct, ifelse(med.fpi.tum > med.fpi.nor, "yellow", "green")) 
  
  fpi.res <- rbind.data.frame(fpi.res,
                              data.frame(sampleID = names(fpi),
                                         fpi = as.numeric(fpi),
                                         tissue = rep(c("Tumor","Normal"), c(length(tumsam),length(norsam))),
                                         p = wt$p.value,
                                         p.lab = ifelse(wt$p.value < 0.001,formatC(wt$p.value,format = "e",digits = 1), round(wt$p.value, 3)),
                                         TCGA = t,
                                         stringsAsFactors = F),
                              stringsAsFactors = F)
}
names(dirct) <- tumors

dirct <- sort(dirct) # 根据标签的颜色排序
fpi.res$TCGA <- factor(fpi.res$TCGA, levels = names(dirct)) # 设置因子
head(fpi.res)
# 输出到文件
write.csv(fpi.res, "output_FPI.csv", quote = F, row.names = F)
```

# 绘制泛癌箱型图

```{r, fig.width=5, fig.height=6}
# 定义颜色
green <- "#2E8049"
yellow <- "#DD8606"

p <- ggplot(fpi.res, aes(TCGA, fpi, fill=tissue)) + 
  geom_boxplot(outlier.shape = NA) + 
  geom_text(aes(TCGA, y = 4,
                label = paste0("P = ", p.lab)),
            data = fpi.res,
            inherit.aes = F) +
  scale_fill_manual(values = c(green, yellow)) + 
  scale_y_continuous(breaks = c(-2.5,0,2.5,5), labels = c(-2.5,0,2.5,5), limits = c(-3,5)) + 
  xlab(NULL) + ylab("FPI") + coord_flip() +
  theme_classic() +
  
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5,vjust = 0.5,size = 11, colour = "black"),
        axis.text.y = element_text(colour = ifelse(as.character(dirct) == "yellow", yellow, green), size = 10), # 若肿瘤FPI大于癌旁，则y轴标签为黄色，否则为绿色
        legend.position = "bottom",
        legend.title = element_blank()) 
p
ggsave("FPI_Pancancer.pdf", width = 5, height = 6)
```

输出的pdf文件是矢量图，可以用Illustrator等矢量图编辑器打开，移动文字位置。

# GEO数据验证FPI

这里所谓的验证，其实就是在另一套数据里看到同样的趋势。

## 输入文件

```{r}
# 数据加载 (数据下载方式可参考FigureYa203ComBat)
gse.expr <- read.table("GSE121689.expr.txt",sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
gse.sinfo <- read.table("GSE121689.sinfo.txt",sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
dmso.sam <- rownames(gse.sinfo)[grep("DMSO",gse.sinfo$Title)] # 取出DMSO样本
erastin.sam <- rownames(gse.sinfo)[grep("erastin",gse.sinfo$Title)] # 取出Erastin样本

# 处理表达谱行名并去重复
gse.expr$genename <- sapply(strsplit(rownames(gse.expr), " /// ", fixed = T), "[", 1)
gse.expr <- gse.expr[!duplicated(gse.expr$genename),]; rownames(gse.expr) <- gse.expr$genename; gse.expr <- gse.expr[,-ncol(gse.expr)] # 去重复基因表达
```

## 计算FPI

```{r}
# ssgsea计算enrichment score
es.gse <- gsva(expr = as.matrix(gse.expr),
               gset.idx.list = fpi.sig,
               method = "ssgsea",
               ssgsea.norm = TRUE)

# 计算FPI
fpi.gse <- as.numeric(scale(es.gse[1,] - es.gse[2,])); names(fpi.gse) <- colnames(gse.expr)
wt <- wilcox.test(fpi.gse[dmso.sam],fpi.gse[erastin.sam]) # 秩和检验
```

## 绘制箱型图

```{r, fig.width=3, fig.height=4}
par(bty="l", mgp = c(2,0.5,0), mar = c(3.1,4.1,2.1,2.1),tcl=-.25,las = 1,xpd = T)
boxplot(fpi.gse[dmso.sam],
        fpi.gse[erastin.sam],
        col = c("#2E8049","#DD8606"),
        ylab = "FPI (GSE121689)",
        xlab = "",
        ylim = c(-2,2), # 控制y轴区间
        names = c("DMSO","Erastin"))
lines(c(1,2),c(2.2,2.2)) # 在两个箱子之间加横线
lines(c(1,1),c(2.1,2.2)) # 垂直于第一个箱子加短竖线
lines(c(2,2),c(2.1,2.2)) # 垂直于第二个箱子加短竖线
text(1.5, 2.3, paste0("p = ",round(wt$p.value, 3))) # 添加p值
dev.copy2pdf(file = "FPI_GSE121689.pdf", width = 3, height = 3.8)
```

# Session Info

```{r}
sessionInfo()
```