---
title: "FigureYa56Immune_inflitrationV3"
author: "小丫画图出品"
date: "2020-2-7"
output: html_document
---
欢迎关注“小丫画图”公众号，回复“小白”，看小视频，实现点鼠标跑代码。

微信ID: epigenomics  E-mail: figureya@126.com

作者：小高

小丫编辑校验

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述

用TCGA的表达数据，计算免疫浸润，画出paper里的heatmap

![](example.png)

出自<https://www.tandfonline.com/doi/full/10.1080/2162402X.2017.1382792?scroll=top&needAccess=true>

# 应用场景

深入挖掘肿瘤RNA-seq数据，跟免疫建立联系。

例文用的是MCPcounter包，输入的表达矩阵可以是microarray，也可以是RNA-seq。TPM、FPKM、normalized read count等等都可以，要取log2。

物种：人，组织（包括血液）的表达数据

参考资料:

- 作者ebecht的回复："The data should be log2-transformed."<https://github.com/ebecht/MCPcounter/issues/4>

- 更多经过测试的基因表达平台，参考这篇：<https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-1070-5>

# 环境设置

安装需要的包

```{r}
#使用国内镜像安装包
options("repos"= c(CRAN="https://mirrors.tuna.tsinghua.edu.cn/CRAN/"))
install.packages("devtools")

library(devtools)
install_github("ebecht/MCPcounter",ref="master", subdir="Source")
```

加载需要用到的包

```{r}
library(MCPcounter)
library(stringr)
library(data.table)
library(pheatmap)

Sys.setenv(LANGUAGE = "en") #显示英文报错信息
options(stringsAsFactors = FALSE) #禁止chr转成factor
```

# 参数设置

```{r}
targetGene <- "CD274" #按哪个基因的表达量排序
#targetCancer <- "glioblastoma multiforme" #提取哪个癌症的表达数据，癌症全称，首字母小写
targetCancer <- "brain lower grade glioma" #LGG
targetMut <- "IDH" #提取哪个基因的突变
```

# 输入数据下载

如果你的数据已经整理成“easy_input_expr.csv”的格式，就跳过这步，直接进入“计算免疫细胞”。

如果你只需要计算免疫细胞，就只需要用到基因表达矩阵：easy_input_expr.csv

如果你要画出paper里那样的图，就还需要临床、突变信息：easy_input_cli.csv

## 文件地址

以下文件已上传到微云：<https://share.weiyun.com/5kthgr4>

从UCSC xena<https://xenabrowser.net/datapages/>进入Pan-Cancer (PANCAN) (39 datasets)

- 点击gene expression RNAseq里的TOIL RSEM tpm (n=10,535) UCSC Toil RNAseq Recompute，下载TPM：<https://toil.xenahubs.net/download/tcga_RSEM_gene_tpm.gz>。

- 点击phenotype里的sample type and primary disease (n=12,804) Pan-Cancer Atlas Hub，下载phenotype：<https://pancanatlas.xenahubs.net/download/TCGA_phenotype_denseDataOnlyDownload.tsv.gz>

- 点击phenotype里的Curated clinical data (n=12,591) Pan-Cancer Atlas Hub，下载带临床信息的生存数据：https://pancanatlas.xenahubs.net/download/Survival_SupplementalTable_S1_20171025_xena_sp.gz

- 点击somatic mutation (SNP and INDEL)里的MC3 public version* (n=9,104) Pan-Cancer Atlas Hub，下载突变：<https://pancanatlas.xenahubs.net/download/mc3.v0.2.8.PUBLIC.xena.gz>

- ID和Gene symbol对应列表下载：<https://toil.xenahubs.net/download/probeMap/gencode.v23.annotation.gene.probemap>

从UCSC xena<https://xenabrowser.net/datapages/>进入TCGA Glioblastoma (GBM) (29 datasets)

- 点击phenotype里的Phenotypes (n=629) TCGA hub，下载带亚型的临床数据：<https://tcga.xenahubs.net/download/TCGA.GBM.sampleMap/GBM_clinicalMatrix.gz>

## 解压缩

*.gz文件是压缩包，解压缩到当前文件夹

## 检查一下：现在你的文件夹里应该有这些输入文件：

tcga_RSEM_gene_tpm

TCGA_phenotype_denseDataOnlyDownload.tsv

Survival_SupplementalTable_S1_20171025_xena_sp

mc3.v0.2.8.PUBLIC.xena

gencode.v23.annotation.gene.probemap

GBM_clinicalMatrix

# 输入数据整理

## 提取目标癌症的sample

```{r}
tcgacase <- read.delim(file="TCGA_phenotype_denseDataOnlyDownload.tsv",header=T,as.is = T)
#查询癌症全称
table(tcgacase$X_primary_disease)

targetCancerCase <- tcgacase[which(tcgacase$X_primary_disease == targetCancer),]
targetCancerCase <- targetCancerCase[targetCancerCase$sample_type %like% "Tumor",] #只留tumor，去掉normal
head(targetCancerCase)
dim(targetCancerCase)
```

## 提取目标癌症的表达矩阵

```{r}
system('head -1 tcga_RSEM_gene_tpm >headtcga')
headtcga <- read.delim("headtcga", as.is=T, check.names = F)

tcgasample <- colnames(headtcga)
targetnum <- which(tcgasample %in% targetCancerCase$sample)
targetnumChr <- str_c(targetnum, collapse = ",")

shellcmd <- paste0("cut -f 1,",targetnumChr," tcga_RSEM_gene_tpm >targetCancerTPM")
system(shellcmd) #大概需要2min

targetCancerTPM <- read.delim("targetCancerTPM", row.names = 1, check.names = F)
targetCancerTPM[1:3, 1:3]
dim(targetCancerTPM)

#有TPM值的目标癌症的sample ID，后面会用到
targetSample <- colnames(targetCancerTPM)
```

TCGA的GBM共有607个sample；

TOIL RSEM tpm流程跑出来的表达数据当中，GBM有166个sample；

GDC来源的表达数据当中，GBM有176个sample。

### 把gene symbol加到最后一列

```{r}
idmap <- read.delim("gencode.v23.annotation.gene.probemap",as.is=T)
head(idmap)
rownames(idmap) <- idmap$id
targetCancerTPM$gsym <- idmap[rownames(targetCancerTPM),]$gene

#此处只保留前2个sample、前1000个基因，用作格式参考
write.csv(targetCancerTPM[1:1000,c(1:2,ncol(targetCancerTPM))],"easy_input_expr_part.csv", quote = F)
#运行下面这行保存全部基因的表达矩阵
#write.csv(targetCancerTPM,"easy_input_expr.csv", quote = F)
```

## 提取目标基因的表达量

```{r}
targetGeneExp0 <- targetCancerTPM[targetCancerTPM$gsym %in% targetGene,]
targetGeneExp0 <- targetGeneExp0[,-ncol(targetGeneExp0)] #删掉最后一列基因名

#按表达量排序
targetGeneExp <- targetGeneExp0[,order(targetGeneExp0)] 
targetGeneExp <- rbind(colnames(targetGeneExp),targetGeneExp)
targetGeneExp <- data.frame(t(targetGeneExp))
colnames(targetGeneExp) <- c("bcr_patient_barcode", targetGene )
head(targetGeneExp)
targetGeneExp$bcr_patient_barcode <- str_replace_all(rownames(targetGeneExp),"[.]","-")
```

## 提取目标癌症的临床信息

Cell paper整理过的生存数据，在嘉因公众号回复“xena”查看详情。

```{r}
clinical <- read.delim("Survival_SupplementalTable_S1_20171025_xena_sp", header = T)
targetCancer_cli<- clinical[clinical$sample %in% targetSample,]
dim(targetCancer_cli)

head(targetCancer_cli)

#查看分期
#table(targetCancer_cli$ajcc_pathologic_tumor_stage)
#table(targetCancer_cli$clinical_stage)
#此处GBM的分期都为空

#有些癌症的分期后面有ABC，想删掉就运行下面这行
#targetCancer_cli$ajcc_pathologic_tumor_stage <- str_replace_all(targetCancer_cli$ajcc_pathologic_tumor_stage,"[ABC]","")

#查看种族
table(targetCancer_cli$race)
#查看性别
table(targetCancer_cli$gender)
```

这些信息都可以画到最后的图上，不同肿瘤情况不一样，自己决定画哪列信息。

## 提取目标癌症的亚型信息

不同癌症的列不一样，GBM有一列GeneExp_Subtype，与例文中的Expression subtype一致。

```{r}
targetCancer_sub <- read.delim("GBM_clinicalMatrix")
targetCancer_sub <- read.delim("LGG_clinicalMatrix")
targetCancer_sub<- targetCancer_sub[targetCancer_sub$sampleID %in% targetSample,]

#table(targetCancer_sub$GeneExp_Subtype) #GBM用这行
table(targetCancer_sub$histological_type) #LGG有组织学分型的列
```

## 提取目标癌症的突变信息

```{r}
mutation <- read.delim("mc3.v0.2.8.PUBLIC.xena", header = T)
targetCancer_mut<- mutation[mutation$sample %in% targetSample,]
all_mutSample <- unique(targetCancer_mut$sample)
length(all_mutSample) #查看该肿瘤共有多少个sample有突变信息

#如果要精确匹配，就把 %like% 换成 ==
targetCancer_mut_gene <- targetCancer_mut[targetCancer_mut$gene %like% targetMut,]
mutSample <- unique(targetCancer_mut_gene$sample) #带有目标突变的sample
#文件中的其他sample被认为是野生型，不在该文件中的sample被认为是未检出，后面将被标为NA
wildSample <- setdiff(all_mutSample, mutSample)
```

## 合并以上多个层面的信息

```{r}
targetAnno <- targetGeneExp
rownames(targetAnno) <- targetGeneExp$bcr_patient_barcode

targetAnno$race <- rep("NA", length(targetSample))
targetAnno$gender <- rep("NA", length(targetSample))
targetAnno$subtype <- rep("NA", length(targetSample))
targetAnno$mut <- rep("NA", length(targetSample))

targetAnno[targetCancer_cli$sample,]$race <- targetCancer_cli$race
targetAnno$race[targetAnno$race == ""] <- "NA" #把人种为空替换成NA

targetAnno[targetCancer_cli$sample,]$gender <- targetCancer_cli$gender
targetAnno$gender[targetAnno$gender == ""] <- "NA" #把性别为空替换成NA

targetAnno[targetCancer_sub$sampleID,]$subtype <- targetCancer_sub$histological_type
#targetAnno[targetCancer_sub$sampleID,]$subtype <- targetCancer_sub$GeneExp_Subtype #GBM用这行
targetAnno$subtype[targetAnno$subtype == ""] <- "NA" #把亚型为空替换成NA

targetAnno[mutSample,]$mut <- "mutant"
targetAnno[wildSample,]$mut <- "wildtype"
table(targetAnno$mut) #查看突变、野生型、NA的数量

head(targetAnno)
write.csv(targetAnno, "easy_input_cli.csv", quote = F, row.names = F)
```

# 计算免疫细胞

计算免疫浸润，只需要全部基因的表达矩阵。

要求:

- 基因ID为HUGO_symbols或ENTREZ_ID。
- 取log2，此处输入文件已经过**log2(x+0.001)**转换。<https://xenabrowser.net/datapages/?dataset=tcga_RSEM_gene_tpm&host=https%3A%2F%2Ftoil.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443>: "Data (file names: *.rsem_genes.results) are downloaded, tpm values are extracted, log2(x+0.001) transformed, and combined."

```{r}
#查看用法
#?MCPcounter.estimate
#MCPcounter需要HUGO_symbols或ENTREZ_ID
#现在第一列是ENSEMBL ID，最后一列是gene symbol
#targetCancerTPM <- read.csv("easy_input_expr.csv", row.names = 1)
dim(targetCancerTPM)

#有多个ENSEMBL ID对应同一个gene symbol
#取中值，如果想取平均值，就把median改为mean
targetCancerTPM_uniq <- aggregate(.~gsym, targetCancerTPM, median) #大概需要5min
dim(targetCancerTPM_uniq)

#现在就可以用gene symbol作行名了
row.names(targetCancerTPM_uniq) <- targetCancerTPM_uniq$gsym
targetCancerTPM_uniq <- targetCancerTPM_uniq[,-1] #删掉最后一列gene symbol
targetCancerTPM_uniq[1:3,1:3]

#计算免疫细胞
targetCancerTPMestimates <- MCPcounter.estimate(targetCancerTPM_uniq, featuresType = "HUGO_symbols")
dim(targetCancerTPMestimates)
write.csv(targetCancerTPMestimates, "MCPcounter_output.csv", quote = F)
```

# 开始画图

```{r}
targetCancerTPMestimates <- read.csv("MCPcounter_output.csv", row.names = 1, check.names = F)
targetAnno <- read.csv("easy_input_cli.csv", row.names = 1)
annotation_col_pre <- targetAnno
annotation_col <- data.frame(annotation_col_pre)
#按照基因表达量排的序
targetCancerTPMestimates <- targetCancerTPMestimates[,row.names(annotation_col)]

pheatmap(targetCancerTPMestimates, 
         method = "spearman", 
         cluster_rows=T, cluster_cols=F,
         cellwidth = 1, cellheight = 30, fontsize = 10, 
         color = colorRampPalette(c("navy","white","red"))(100),
         scale="row",show_colnames = F,
         border_color = "NA",
         annotation_col = annotation_col,
         filename = "MCPcounter.pdf")
```

![](MCPcounter.pdf)

```{r}
sessionInfo()
```