---
title: "FigureYa110mutationSignature"
author: "小丫画图出品"
date: "2019-6-27"
output: html_document
---
欢迎关注“小丫画图”公众号，同名知识星球等你加入

小丫微信: epigenomics  E-mail: figureya@126.com

作者：大鱼海棠

单位：Research Center of Biostatistics and Computational Pharmacy, China Pharmaceutical University

小丫编辑校验

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 需求描述

画出文章里的3个热图：

![](example.png)

出自<https://academic.oup.com/annonc/article/28/7/1597/3611460>

## 应用场景

利用deconstructsigs计算突变签名模式，NMF做聚类分析，寻找亚型驱动签名，并绘制热图。

**注意：若使用本众筹分析流程，请引用：**

Alexandrov LB, Nik-Zainal S, Wedge DC, Aparicio SA, Behjati S, Biankin AV, et al. Signatures of mutational processes in human cancer. Nature 2013;500(7463):415.

Rosenthal R, McGranahan N, Herrero J, Taylor BS, Swanton C. DeconstructSigs: delineating mutational processes in single tumors distinguishes DNA repair deficiencies and patterns of carcinoma evolution. Genome biology 2016;17(1):31.

Gaujoux R , Seoighe C . A flexible R package for nonnegative matrix factorization[J]. BMC Bioinformatics, 2010, 11(1):367.

## 环境设置

使用国内镜像安装包

```r
options("repos"= c(CRAN="https://mirrors.tuna.tsinghua.edu.cn/CRAN/"))
options(BioC_mirror="http://mirrors.ustc.edu.cn/bioc/")
BiocManager::install("BSgenome")
BiocManager::install("BSgenome.Hsapiens.UCSC.hg19")
BiocManager::install("GenomeInfoDb")
```

deconstructsigs的安装：从github下载<https://github.com/raerose01/deconstructSigs>，解压缩，然后压缩为.tar.gz格式，本地安装。

```bash
tar -czvf deconstructSigs.tar.gz deconstructSigs-master
```

加载包

```{r}
library(tidyverse)
library(magrittr)
library(readxl)
library(stringr)
library(forcats)
library(deconstructSigs)
library(BSgenome.Hsapiens.UCSC.hg19)
library(NMF)

Sys.setenv(LANGUAGE = "en") #显示英文报错信息
options(stringsAsFactors = FALSE) #禁止chr转成factor
```

## 输入文件

blca.mut.maf.txt，基因突变数据。

```{r}
### 读取BLCA突变数据 ###
maf <- read_tsv("blca.mut.maf.txt", comment = "#")
maf <- as.data.frame(maf)
```

## 开始画图

```{r}
fig.path <- "Figures" #每个样本的signature细节图保存在Figures文件夹里
if (!file.exists(fig.path)) { dir.create(fig.path) }
#设置颜色
jco <- c("#2874C5","#EABF00","#C6524A","#868686")
```

### 计算96种三核苷酸变化 ###

```{r}
rmSilence = F #是否移除沉默突变（根据实际情况还可能移除其他SNP类型）;移除则改为T

if (rmSilence) {
  maf <- as.data.frame(maf[which(maf$Variant_Type == "SNP" & maf$Variant_Classification != "Silent"),]) #仅考虑SNP并移除沉默突变
} else {
  maf <- as.data.frame(maf[which(maf$Variant_Type == "SNP"),]) #仅使用SNP (突变签名研究本身只考虑SNP)
}
maf$Chromosome <- paste0("chr",maf$Chromosome) # 保证染色体格式为chrx，符合deconstructsigs输入需求
snp.count <- mut.to.sigs.input(mut.ref = maf, 
                                sample.id = "Tumor_Sample_Barcode", 
                                chr = "Chromosome", 
                                pos = "Start_Position", 
                                ref = "Reference_Allele", # 参考位点碱基
                                alt = "Tumor_Seq_Allele2", # 突变位点碱基
                                bsg = BSgenome.Hsapiens.UCSC.hg19) # hg19参考基因组
write.table(snp.count,"snp.count.txt",sep = "\t",row.names = T,col.names = NA)
```

### 计算单样本突变签名的权重 ###

```{r}
cut.off <- 0.06 # 默认cutoff为6%，如例文所示
mut.wt <- data.frame()
sigs.out.list <- list()
index <- 1
for (sample in rownames(snp.count)) {
  cat(paste0(sample," starts and ",length(rownames(snp.count))-index," samples remain to be analyzed!\n"))
  
  tmp <- whichSignatures(tumor.ref = snp.count, 
                         signatures.ref = signatures.cosmic, 
                         sample.id = sample, 
                         contexts.needed = TRUE,
                         tri.counts.method = 'exome2genome', # 标准化方式如例文所示
                         signature.cutoff = cut.off)
  
  index <- index + 1
  # 输出每个样本的签名细节图
  pdf(file.path(fig.path,paste0(sample,"_plotSignatures.pdf")))
  plotSignatures(tmp)
  invisible(dev.off())
  
  pdf(file.path(fig.path,paste0(sample,"_weightPie.pdf")))
  makePie(tmp)
  invisible(dev.off())
  
  # 生成突变签名矩阵
  sigs.out.list[[sample]] <- tmp
  tmp <- data.frame(c(tmp$weights,unknown=tmp$unknown),row.names = sample)
  mut.wt <- rbind.data.frame(mut.wt,tmp)
}
write.table(mut.wt,"mutsig.weightMatrix.txt",sep = "\t",row.names = T,col.names = NA)
```

### 非负矩阵分解识别驱动性signature并绘图 ###

```{r}
#搜索最优rank（不宜过大）
nmf.input <- t(mut.wt)
nmf.input <- nmf.input[setdiff(rownames(nmf.input),"unknown"),] #去掉unknown
ranks <- 2:5
estim <- lapply(ranks, function(r){
  fit <- nmf(nmf.input, r, nrun = 5, seed = 4, method = "lee") #nrun设置为5以免运行时间过长
  list(fit = fit, consensus = consensus(fit), .opt = "vp",coph = cophcor(fit))
})
names(estim) <- paste('rank', ranks)

# 绘制随rank变化的cophenetic系数图
pdf("Cophenetic coefficient for seleting optimal nmf rank.pdf")
par(cex.axis=1.5)
plot(ranks, sapply(estim, '[[', 'coph'), xlab="", ylab="", type="b", col="red", lwd=4,xaxt="n")
axis(side = 1, at=1:5)
title(xlab="number of clusters", ylab="Cophenetic coefficient", cex.lab=1.5)
invisible(dev.off())

# 根据cophenetic得分选择rank=4
rank <- 4
seed <- 2019620
rownames(nmf.input) <- gsub("Signature","Sig",rownames(nmf.input)) # 行名简化
mut.nmf <- nmf(nmf.input, 
               rank = rank, 
               seed = seed, 
               method = "lee") 

# 注意原文是先利用每个样本中signature的占比来筛选“显著频发”的突变
# 这里众筹采用NMF自身提取亚型驱动因子的方式寻找显著突变
index <- extractFeatures(mut.nmf,"max") # 理论上每个亚型都会有特征，也不排除某亚型没有特征的情况，输出值为NA，若更改方法请参考??extractFeatures
sig.order <- unlist(index)

# 使用挑选出的signature再次NMF
nmf.input2 <- nmf.input[sig.order,]
mut.nmf2 <- nmf(nmf.input2, 
                rank = rank, 
                seed = seed, 
                method = "lee") 
group <- predict(mut.nmf2) # 提出亚型

# 样本顺序，因为nmf没有树结构，所以最后画热图需要按照类排序，这里将按照类名1，2，3，4升序排列
sample.order <- names(group[order(group)])

# consensus heatmap
pdf(file = "consensusmap.pdf",width = 4,height = 4)
consensusmap(mut.nmf2,
             labRow = NA,
             labCol = NA,
             annCol = data.frame("cluster"=group[colnames(nmf.input)]),
             annColors = list(cluster=c("1"=jco[1],"2"=jco[2],"3"=jco[3],"4"=jco[4])))
invisible(dev.off())

# basismap是aheatmap的改版，请参照aheatmap参数调试
pdf(file = "basismap.pdf",width = 4.5,height = 4) # 从这张图可以比较清晰地看到各亚型中的驱动signature（深色），与下面的nmf heatmap对应
basismap(mut.nmf2,
         cexCol = 1,
         cexRow = 0.3,
         annColors=list(c("1"=jco[1],"2"=jco[2],"3"=jco[3],"4"=jco[4])))
invisible(dev.off())

# nmf heatmap
# 既然用了人家的NMF，那就直接采用NMF自带的aheatmap绘图^^
aheatmap(as.matrix(nmf.input2[,sample.order]), 
         Rowv=NA, 
         Colv=NA, 
         annCol = data.frame("cluster"=group[sample.order]),
         annColors = list(cluster=c("1"=jco[1],"2"=jco[2],"3"=jco[3],"4"=jco[4])),
         color=c("#EAF0FA","#6081C3","#3454A7"), # 例文的蓝色渐变
         revC=TRUE, 
         cexCol = 0.3, 
         cexRow = 0.3,
         filename = "NMF_heatmap.pdf")
```

```{r}
sessionInfo()
```