---
title: "FigureYa55panCancer_violinV2"
author: "小丫画图出品"
date: "2019-1-29"
output: html_document
---
微信ID: epigenomics  E-mail: figureya@126.com

GTEx和TCGA数据处理：小丫、徐洲更

violin图：Byron

小丫编辑校验

## 需求描述

TCGA里的正常组织太少，把GTEx的正常组织加进去，用分裂小提琴图展示某个基因在TCGA肿瘤和正常组织中的表达差异。

![](example.png)

出自<https://www.sciencedirect.com/science/article/pii/S0896627318308481>, Supplemental material, Figure 3b

## 应用场景

适用于：

1. 对比某一基因在TCGA肿瘤组织和正常组织(TCGA的normal和GTEx)基因表达量TPM值

2. 用分裂小提琴图展示表达量数据

## 环境设置

安装需要用到的包

```r
#使用国内镜像安装包
options("repos"= c(CRAN="https://mirrors.tuna.tsinghua.edu.cn/CRAN/"))
options(BioC_mirror="http://mirrors.ustc.edu.cn/bioc/")
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("TCGAbiolinks", version = "3.8")

install.packages("stringr")
install.packages("dplyr")
install.packages("ggplot2")
install.packages("RColorBrewer")
```

加载需要用到的包

```{r}
library(stringr)
library(dplyr)
library(ggplot2)
library(RColorBrewer)
source("read_part.R") # autor: 徐洲更
Sys.setenv(LANGUAGE = "en") #显示英文报错信息
options(stringsAsFactors = FALSE) #禁止chr转成factor
```

## 参数设置

```{r}
targetGene <- "TP53" # 想要画的目标基因
```

## 输入文件下载

如果你的数据已经整理成easy_input.csv的格式，就可以跳过这步，直接进入“输入数据预处理”。

### 文件地址：

从UCSC xena<https://xenabrowser.net/datapages/>下载经过TOIL流程统一处理的TCGA和GTEx的TPM，free of computational batch effects。

参考资料：The goal of the Toil recompute was to process ~20,000 RNA-seq samples to create a **consistent meta-analysis** of four datasets **free of computational batch effects**. <https://xenabrowser.net/datapages/?hub=https://toil.xenahubs.net:443>

用这套数据画出来的图跟xena的violin一致：<https://xenabrowser.net/transcripts/>

- TCGA Pan-Cancer (PANCAN) (39 datasets)，点击gene expression RNAseq里的TOIL RSEM tpm (n=10,535) UCSC Toil RNAseq Recompute，下载TPM：<https://toil.xenahubs.net/download/tcga_RSEM_gene_tpm.gz>。点击phenotype里的sample type and primary disease (n=12,804) Pan-Cancer Atlas Hub，下载phenotype：<https://pancanatlas.xenahubs.net/download/TCGA_phenotype_denseDataOnlyDownload.tsv.gz>

- GTEX (11 datasets)，点击gene expression RNAseq里的TOIL RSEM tpm (n=7,862) UCSC Toil RNAseq Recompute，下载TPM：<https://toil.xenahubs.net/download/gtex_RSEM_gene_tpm.gz>。点击phenotype里的GTEX phenotype (n=9,783) UCSC Toil RNAseq Recompute，下载phenotype：<https://toil.xenahubs.net/download/GTEX_phenotype.gz>

- ID和Gene symbol对应列表下载：<https://toil.xenahubs.net/download/probeMap/gencode.v23.annotation.gene.probemap>

- TCGA跟GTEx的组织对应关系，参照GEPIA help的Differential analysis：<http://gepia.cancer-pku.cn/help.html>，整理成samplepair.txt文件

**题外话：**

- 在<https://xenabrowser.net/datapages/?cohort=TCGA%20Pan-Cancer%20(PANCAN)&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443>这个页面，还有基于TCGA数据加工过的iCluster、Molecular subtype等宝贝等你挖掘。

- 在<https://xenabrowser.net/datapages/>目录下面，有TCGA的各种测序数据，下载链接有规律可循，可批量下载。

### 检查一下：你的文件夹里应该有这些输入文件：

tcga_RSEM_gene_tpm.gz

TCGA_phenotype_denseDataOnlyDownload.tsv.gz

gtex_RSEM_gene_tpm.gz

GTEX_phenotype.gz

gencode.v23.annotation.gene.probemap

samplepair.txt

## 提取目的基因在各种肿瘤和正常组织里的TPM

### 为GTEx的sample标注组织

```{r}
samplepair <- read.delim("samplepair.txt",as.is = T)
gtexpair <- samplepair[,c(1,2,5)]
gtexpair

#以TCGA的缩写为GTEx的组织命名
gtexpair$type <- paste0(gtexpair$TCGA,"_normal_GTEx")
gtexpair$type2 <-"normal"
gtextcga <- gtexpair[,c(1,3:5)] #筛掉Detail列
colnames(gtextcga)[1:2] <- c("tissue","X_primary_site")
head(gtextcga)

#为GTEx的sample标出组织
gtexcase <- read.delim(file="GTEX_phenotype.gz",header=T,as.is = T)
colnames(gtexcase)[1] <- "sample"
gtexcase2tcga <- merge(gtextcga,gtexcase,by="X_primary_site")
gtextable <- gtexcase2tcga[,c(5,2:4)]
head(gtextable)
```

### 为TCGA的sample标注组织

```{r}
tissue <- gtexpair$TCGA
names(tissue) <- gtexpair$Detail

tcgacase <- read.delim(file="TCGA_phenotype_denseDataOnlyDownload.tsv.gz",header=T,as.is = T)

tcgacase$tissue <- tissue[tcgacase$X_primary_disease]
tcgacase$type <- ifelse(tcgacase$sample_type=='Solid Tissue Normal',paste(tcgacase$tissue,"normal_TCGA",sep="_"),paste(tcgacase$tissue,"tumor_TCGA",sep="_"))
tcgacase$type2 <- ifelse(tcgacase$sample_type=='Solid Tissue Normal',"normal","tumor")
tcgatable<-tcgacase[,c(1,5:7)]
head(tcgatable)
```

### 删除没有TPM的sample

```{r}
headgtex <- read_part("gtex_RSEM_gene_tpm.gz", rows = 1)
gtexsample <- headgtex[1,]
gtextable2 <- gtextable[gtextable$sample %in% gtexsample,]

headtcga <- read_part("tcga_RSEM_gene_tpm.gz", rows = 1)
tcgasample <- headtcga[1,]
tcgatable2 <- tcgatable[tcgatable$sample %in% tcgasample,]

tcga_gtex<-rbind(tcgatable2,gtextable2)
```

### 比较现有数据跟GEPIA收录的sample数量

```{r}
statable<-data.frame()
for (tissuer in sort(unique(tcga_gtex$tissue))){
  
  tcga_gtex_tissuer<-tcga_gtex[tcga_gtex$tissue==tissuer,]

  tissuer_tumor_TCGA <- paste0(tissuer,"_tumor_TCGA")
  tissuer_normal_TCGA <- paste0(tissuer,"_normal_TCGA")
  tissuer_normal_GTEx <- paste0(tissuer,"_normal_GTEx")

  tissuerdf<-data.frame(TCGA=tissuer,
                     TCGA_new_tumor=sum(tcga_gtex_tissuer$type==tissuer_tumor_TCGA),
                     TCGA_new_normal=sum(tcga_gtex_tissuer$type==tissuer_normal_TCGA),
                     GTEx_new_num=sum(tcga_gtex_tissuer$type==tissuer_normal_GTEx),
                     stringsAsFactors=F)
  #str(tissuerdf)
  statable<-rbind(statable,tissuerdf)
}
cbind(statable, samplepair[,c(1,3:6)])
```

前4列是现有数据，后面是GEPIA数据

### 找出每个基因在TPM文件中的行号

```{r}
idmap <- read.delim("gencode.v23.annotation.gene.probemap",as.is=T)
head(idmap)
# 读取第一列, 所有行
tpmid2row <- read_part("gtex_RSEM_gene_tpm.gz", rows = -1, columns = 1)
tpmid2row <- tpmid2row[2:nrow(tpmid2row)]
tpmid2row <- data.frame(id = tpmid2row)
tpmid2row$rownum <- seq(1,nrow(tpmid2row))
head(tpmid2row)

gene2id2row <- merge(idmap,tpmid2row,by="id")
gene2id2row <- gene2id2row[order(gene2id2row$rownum),]
head(gene2id2row)
```

### 提取目的基因在肿瘤和正常组织的TPM

此处以TP53基因为例，提取它的TPM

```{r}
generownum <- gene2id2row[gene2id2row$gene==targetGene,]$rownum + 1
gtex_colnames <- read_part("gtex_RSEM_gene_tpm.gz", rows = 1)
gtex_tpm <- read_part("gtex_RSEM_gene_tpm.gz", rows = generownum)
tcga_colnames <- read_part("tcga_RSEM_gene_tpm.gz", rows = 1)
tcga_tpm <- read_part("tcga_RSEM_gene_tpm.gz", rows = generownum)

tpm <- as.numeric(c(gtex_tpm[2:length(gtex_tpm)], 
                    tcga_tpm[2:length(tcga_tpm)] ))
names(tpm) <- c(gtex_colnames[2:length(gtex_colnames)], 
                    tcga_colnames[2:length(tcga_colnames)] )

tcga_gtex$tpm <- tpm[tcga_gtex$sample]

#按组织排序
tcga_gtex <- arrange(tcga_gtex,tissue)
write.csv(tcga_gtex[,c(2,4,5)],"easy_input.csv", quote = F)
```

## 输入数据预处理

easy_input.csv：第1列tissue是组织（分组），第2列type2是肿瘤/正常组织（左右分组），第3列tpm是TPM值。

有些组织没有normal，需要识别出来单独画。

```{r}
tcga_gtex <- read.csv("easy_input.csv", row.names = 1, header = T, as.is = F)
head(tcga_gtex)

tumorlist <- unique(tcga_gtex[tcga_gtex$type2=="tumor",]$tissue)
normallist <- unique(tcga_gtex[tcga_gtex$type2=="normal",]$tissue)
withoutNormal <- setdiff(tumorlist, normallist)
```

MESO和UVM没有normal

```{r}
tcga_gtex$type2 <- factor(tcga_gtex$type2,levels=c("tumor","normal"))
tcga_gtex_withNormal <- tcga_gtex[!(tcga_gtex$tissue %in% withoutNormal),]
tcga_gtex_MESO <- tcga_gtex[tcga_gtex$tissue=="MESO",]
tcga_gtex_UVM <- tcga_gtex[tcga_gtex$tissue=="UVM",]
```

## 开始画图

要用到SplitViolin的图层和函数，保存在GeomSplitViolin.R文件中，位于当前文件夹。

```{r, fig.width=20, fig.height=6}
source("GeomSplitViolin.R") #位于当前文件夹

p <- ggplot(tcga_gtex_withNormal, aes(x = tissue, y = tpm, fill = type2)) + #x对应肿瘤的类型，y对应表达量，fill填充对应组织的类型
  geom_split_violin(draw_quantiles = c(0.25, 0.5, 0.75), #画4分位线
                    trim = F, #是否修剪小提琴图的密度曲线
                    linetype = "solid", #周围线的轮廓
                    color = "black", #周围线颜色
                    size = 0.2,
                    na.rm = T,
                    position ="identity")+ #周围线粗细
  ylab(paste0(targetGene, " expression (log2TPM)")) + xlab("") +
  ylim(-4,9) +
  scale_fill_manual(values = c("#DF2020", "#DDDF21"))+
  theme_set(theme_set(theme_classic(base_size=20)))+
  theme(axis.text.x = element_text(angle = 45, hjust = .5, vjust = .5)) + #x轴label倾斜45度
  
  guides(fill = guide_legend(title = NULL)) + 
  theme(legend.background = element_blank(), #移除整体边框
        #图例的左下角置于绘图区域的左下角
        legend.position=c(0,0),legend.justification = c(0,0))

p + geom_split_violin(data = tcga_gtex_MESO,
                       mapping = aes(x = tissue, y = tpm, fill = type2),
                       draw_quantiles = c(0.25, 0.5, 0.75), #画4分位线
                    trim = F, #是否修剪小提琴图的密度曲线
                    linetype = "solid", #周围线的轮廓
                    color = "black", #周围线颜色
                    size = 0.2,
                    na.rm = T,
                    position ="identity") +
  geom_split_violin(data = tcga_gtex_UVM,
                       mapping = aes(x = tissue, y = tpm, fill = type2),
                       draw_quantiles = c(0.25, 0.5, 0.75), #画4分位线
                    trim = F, #是否修剪小提琴图的密度曲线
                    linetype = "solid", #周围线的轮廓
                    color = "black", #周围线颜色
                    size = 0.2,
                    na.rm = T,
                    position ="identity") +
  scale_x_discrete(limits = levels(tcga_gtex$tissue))

ggsave(paste0(targetGene, "expression.pdf"),width = 20, height = 6)
```

```{r}
sessionInfo()
```

