---
title: "FigureYa67phastCons"
author: "小丫画图出品"
date: "2019-1-13"
output: html_document
---
欢迎关注“小丫画图”公众号，同名知识星球等你加入

小丫微信: epigenomics  E-mail: figureya@126.com

作者：王童鞋

小丫编辑校验

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 需求描述

输入ucsc上的phastCons文件、ChIP-seq的peak bed文件。用R或python计算数值，用R画出conservation profile

![](example.png)

出自<https://www.nature.com/articles/ncomms7315>

## 应用场景

展示特定区域的保守性。

重要区域保守性往往较高，例如转录因子结合位点。

## 环境设置

操作系统：Linux或MAC。

需要用到awk、python2.7

下载bigWigSummary：

```r
## choose by platform & cpu type
## macOSX
#download.file("http://hgdownload.soe.ucsc.edu/admin/exe/macOSX.x86_64/bigWigSummary","bigWigSummary")
## linux x86_64
#download.file("http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/bigWigSummary","bigWigSummary")
## linux x86_64 v369
#download.file("http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64.v369/bigWigSummary","bigWigSummary")

#Sys.chmod(c("bigWigSummary"),755)
```

## 输入文件

此处对比all dsDMR、non exon dsDMR和随机序列这三个bed的phastCons

- all_dsDMR.bed, non_exon_dsDMR.bed, download from <https://epigenome.wustl.edu/Zebrafish_DNAme/browse.php>

- random.bed, random regions generation (randomBed -l 500 -n 8225 -seed 6666 -g danRer7.chrom.sizes > random.bed)

- fish.phastCons8way.bw，保守性文件，下载的步骤：

Step1.进入物种选择页面，<http://genome.ucsc.edu/>

![](Step1.png)

Step2.选择物种和基因组版本

![](Step2.png)

Step3.找到保守性记录

![](Step3.png)

Step4.找到保守性文件存放文件夹

![](Step4.png)

Step5.找到保守性文件

![](Step5.png)

Step6.下载保守性文件

![](Step6.png)

## 参数设置

```{r}
outputName <- "phastCons_example.pdf" #输出的文件名
bigWigFile <- c("fish.phastCons8way.bw") #前面下载的保守性文件
bedFiles <- c("all_dsDMR.bed","non_exon_dsDMR.bed","random.bed") #输入的想要对比的bed文件
labels <- c("8,225 dsDMRs","7,965 dsDMRs (non-exonic)","8,225 random regions") #图例里标注的内容
siteType <- "dsDMR centre" #X轴title，Distance from dsDMR centre，根据自己数据来修改
resolution <- 50 #分辨率
span <- 5000 #左右跨度
coreNumber <- 3 #内核数量
colors <- c("black", "indianred3", "gray") #曲线颜色
captureMethod <- "security" 
```

## 提取信号

### 先写个函数，函数里面会调用getBigWigValue.py，位于当前文件夹

```{r}
regionCapture <- function(seqname,start,end,bw,datapoints,captureMethod="security",cores=4L) {
  # capture bigwig signal in region with given datapoints
  captureRegion <- data.frame(seqname=seqname,start=start,end=end)
  write.table(captureRegion,file="captureRegion.bed",quote=F,sep="\t",row.names=F,col.names=F)
  
  # 这里用到了python脚本
  suppressMessages(system(paste("python getBigWigValue.py -b", "captureRegion.bed", "-w", bw, "-n regionCapture", "-p", cores,"-s", datapoints, "-m", captureMethod)))
  signal <- read.table(gzfile("signal_regionCapture_siteprof1.gz"))
  
  return(signal)
}

signal_caputer_around_sites <- function(bigWigFile,bed,resolution=50,span=3000,captureMethod="security",cores=4L) {
  # options
  # bigWigFile: bigWigFile to be capture
  # bedFile: bedFile contain sites information
  # resolution: resolution for capture
  # span: span for upstream, downstream & gene body normalization length
  # cores: cores used for parallel
  
  datapoints <- span * 2 / resolution + 1
  
  seqname <- bed[,1]
  midpoints <- as.integer((bed[,2] + bed[,3] - 1) / 2)
  # get strand information
  if(ncol(bed)>5) { 
    strand <- bed[,6]
  } else {
    strand <- rep(".",nrow(bed))
  }
  
  # capture
  # IRange is 1-based
  signal <- regionCapture(seqname,midpoints-span-resolution/2,midpoints+span+resolution/2,bigWigFile,datapoints,captureMethod,cores)
  
  # correct orientation & get average signal
  write.table(signal,file="temp_signal.txt",quote=F,row.names=F,col.names=F,sep="\t")
  write.table(strand,file="temp_strand.txt",quote=F,row.names=F,col.names=F,sep="\t")
  system(paste0("paste temp_signal.txt temp_strand.txt | awk '{if($NF==",'"',"+",'"',") {for(i=1;i<NF-1;i++) printf $i",'"',"\t",'"',"; print $(NF-1)} else {for(i=NF-1;i>1;i--) printf $i",'"',"\t",'"',"; print $1}}' > temp_signal_rev.txt"))
  signal_rev <- read.table("temp_signal_rev.txt")
  delete_result <- file.remove("temp_signal.txt")
  delete_result <- file.remove("temp_strand.txt")
  delete_result <- file.remove("temp_signal_rev.txt")
  return(apply(signal_rev,2,mean))
  
}
```

### 提取信号

```{r}
average_signal <- c()
for(i in 1:length(bedFiles)){
  
  # load bed file
  bed <- read.table(bedFiles[i])
  isNormalChrosome <- !grepl("_",bed$V1)
  bed <- bed[isNormalChrosome,]
  
  # caputre signal
  average_signal <- rbind(average_signal, signal_caputer_around_sites(bigWigFile, bed, resolution=resolution, span=span, captureMethod=captureMethod, cores=as.integer(coreNumber)))
}

# 需要运行约5min，把信号保存到文件，便于后期重新画图
write.csv(average_signal, "average_signal.csv", quote = F, row.names = F)
```

## 开始画图

```{r}
average_signal <- read.csv("average_signal.csv")
head(average_signal[,c(1:3,(ncol(average_signal)-2):ncol(average_signal))])

# y轴范围
minimum <- min(average_signal) - (max(average_signal) - min(average_signal)) * 0.1
maximum <- max(average_signal) + (max(average_signal) - min(average_signal)) * 0.1 * length(bedFiles)

datapoints <- span / resolution

#pdf(outputName, width = 6, height = 5)
# 画第1条线
plot(1:(datapoints*2+1), average_signal[1,], type="l", lwd = 2, col=colors[1], 
     #bty="n", 
     xaxs="i", yaxs="i", xaxt="n", 
     xlab=paste("Distance from", siteType), ylab="Average PhastCons score", 
     ylim=c(minimum, maximum))

# 画第2、3条线
for(i in 2:length(bedFiles)){
  lines(1:(datapoints*2+1), lwd = 2, average_signal[i,], col=colors[i])
}

# 图例
legend("topright", col=colors, legend=labels, lty=1, lwd = 2, bty="n")

# 坐标轴
axis(side=1, 
     at=c(0, datapoints, datapoints * 2) + 1, 
     labels = c(paste(-1*span/1000,"kb"), "0", paste(span/1000,"kb")))

#dev.off()
```

```{r}
sessionInfo()
```