---
title: "FigureYa147interaction"
author: "小丫画图出品"
date: "2019-11-17"
output: html_document
---
欢迎关注“小丫画图”公众号，同名知识星球等你加入

小丫微信: epigenomics  E-mail: figureya@126.com

作者：Hazard

小丫编辑校验

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 需求描述

用interaction文件作为输入画调控网络图，线的长短与相互作用score对应，线的数量和颜色与证据对应。

![](example1.png)

出自<https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4951236>

这篇文章用matlab写了个脚本OncoBinder.m，free online，能根据多种数据找interaction、打分、画图，感兴趣的小伙伴写信给文章作者吧～

![](example2.png)

出自<https://www.nature.com/articles/s41589-018-0161-x>

Fig. 1 | Identification of HIP1R as a regulator of PD-L1. a, Schematic representation of the identification process. A set of low-confidence interactors of PD-L1 according to high-throughput assays (left panel) were analyzed by the OncoBinder model (middle panel) to identify functionally relevant interactions. The ‘co-activation’ scores of candidate interactors are plotted in the right panel, with statistically significant associations marked in red (P < 0.01, Pearson correlation, two-sided). The modeling experiment was replicated for three times with the same results obtained.

**图的解析**

用改写的OncoBinder（没有提供脚本）找interaction并画图。中央是PD-L1，周围是跟它有相互作用关系的蛋白质。

Figure legend和正文都没有描述线的长短和数量与哪个特征相对应。

我们结合这两幅图设计了下面的图，设计思路看这篇：<https://mp.weixin.qq.com/s/cQzsw1DKTUqooBmE078k2A>

## 应用场景

展示相互作用关系，同时展示相互作用证据的来源和打分。

像例文中的蛋白相互作用，或转录因子-靶基因，或miRNA-靶基因的关系。

其中转录因子-靶基因的关系可以这样设计：中央画转录因子，周围画靶基因；连线代表各种证据例如有motif、有ATAC-seq的peak、有H3K27ac的peak等等；自己给这些证据权重，计算一个总体的打分作为连线的长度。

## 环境设置

使用国内镜像安装包

```{r}
options("repos"= c(CRAN="https://mirrors.tuna.tsinghua.edu.cn/CRAN/"))
options(BioC_mirror="http://mirrors.ustc.edu.cn/bioc/")
```

加载包

```{r}
library(org.Hs.eg.db)
library(clusterProfiler)
library(ggplot2)
library(reshape2)
library(igraph)
library(magrittr)

Sys.setenv(LANGUAGE = "en") #显示英文报错信息
options(stringsAsFactors = FALSE) #禁止chr转成factor
```

自定义画图函数

`fcircle` is built for the white background of PD-L1 

```{r}
mycircle <- function(coords, v=NULL, params) {
  vertex.color <- params("vertex", "color")
  if (length(vertex.color) != 1 && !is.null(v)) {
    vertex.color <- vertex.color[v]
  }
  vertex.size  <- 1/200 * params("vertex", "size")
  if (length(vertex.size) != 1 && !is.null(v)) {
    vertex.size <- vertex.size[v]
  }
  vertex.frame.color <- params("vertex", "frame.color")
  if (length(vertex.frame.color) != 1 && !is.null(v)) {
    vertex.frame.color <- vertex.frame.color[v]
  }
  vertex.frame.width <- params("vertex", "frame.width")
  if (length(vertex.frame.width) != 1 && !is.null(v)) {
    vertex.frame.width <- vertex.frame.width[v]
  }
  
  mapply(coords[,1], coords[,2], vertex.color, vertex.frame.color,
         vertex.size, vertex.frame.width,
         FUN=function(x, y, bg, fg, size, lwd) {
           symbols(x=x, y=y, bg=bg, fg=fg, lwd=lwd,
                   circles=size, lheight = 30, add=TRUE, inches=FALSE)
         })
}

add.vertex.shape("fcircle", clip=igraph.shape.noclip,
		 plot=mycircle, parameters=list(vertex.frame.color=1,
                                  vertex.frame.width=1))
```

## 参数设置

```{r}
# Give your focus official gene_symbol
sourceGene <- "CD274"

# Given name will be shown on the plot instead.
sourceName <- "PD-L1" 
```

## 输入文件

如果你已经获得interaction关系，就把自己的数据整理成easy_input.csv文件的格式，跳过这步，直接进入“开始画图”。

### 相互作用文件的下载

- 数据库。这里用ComPPI数据库检索interaction：Compartmentalized Protein-Protein Interaction Database | v2.1.1, <http://comppi.linkgroup.hu/protein_search>，适用于人、果蝇、线虫和酵母。

![](step1.png)

其他物种可以用STRING检索，也有score。找更复杂的相互作用关系可参考FigureYa53PPI，用cytoscape画图可参考FigureYa54cytoscape。

- 检索蛋白质"CD274"

![](step2.png)

- 下载interaction文件。As the result of your search query, the following two proteins were found.    
Q9NZQ7 (UniProtKB-AC)	Programmed cell death 1 ligand 1	   
Q0GN75 (UniProtKB-AC)	CD274   

I chose Q0GN75 for use. 

Download file "comppi--interactors_of_Q9NZQ7.txt".
<http://comppi.linkgroup.hu/protein_search/interactors/Q0GN75/download>

![](step3.png)

### 输入数据的整理

comppi--interactors_of_Q0GN75.txt，上一步获得的interaction信息。每行一个蛋白质，所有证据和分数都在Major.Loc.With.Loc.Score里面，用｜分隔，需要拆开成每条证据单独一行。

```{r}
idata <- read.table(file = "comppi--interactors_of_Q0GN75.txt", sep = "\t", header = T, quote = "")
idata[1:4,]

# 提取各种证据及其对应的score，每个证据单独一行
ilinks <- NULL
for(i in 1:nrow(idata)){
  for(j in strsplit(idata$Major.Loc.With.Loc.Score[i], split = "\\|")[[1]]){
    tmp <- strsplit(j, split = "\\(")[[1]]
    ilinks <- rbind(ilinks, c(idata$Interactor[i], 
                              tmp[1],
                              gsub(pattern = "\\)", "", tmp[2])))
  }
}
colnames(ilinks) <- c("Interactor", "Loc","LocScore")
ilinks <- as.data.frame(ilinks)
head(ilinks)

# filter out outlier link
# 这里提取以下6种，实际应用时根据你自己的数据修改
ilinks <- ilinks[ilinks$Loc %in% c("Cytosol", "Secretory-pathway", "Extracellular", "Membrane","Mitochondrion", "Nucleus"),]

# convert uniprot id to gene symbol
idMap <- bitr(idata$Interactor, "UNIPROT", toType = "SYMBOL", OrgDb = org.Hs.eg.db)
ilinks$gene <- idMap$SYMBOL[match(ilinks$Interactor, idMap$UNIPROT)]

# 根据网页提供的公式计算相互作用打分：
# http://comppi.linkgroup.hu/help/scores
ilinks$LocScore <- as.numeric(ilinks$LocScore)
locScoreMat <- dcast(ilinks, formula = gene ~Loc, value.var = "LocScore", fun.aggregate = mean)
locScoreMat[is.na(locScoreMat)] <- 0
rownames(locScoreMat) <- locScoreMat$gene
locScoreMat <- locScoreMat[,-1]
locScoreMat$InteractionScore <- 1- apply(locScoreMat[,1:6], 1, function(x) prod(1-x*locScoreMat[sourceGene,]))

# prepare network elements
ilinks$source <- sourceGene
ilinks$target <- ilinks$gene
links <- unique(ilinks[,c("source", "target", "Loc")]) # remove duplicated genes
links <- links[links$target  != sourceGene,] # remove sourceGene 
links$score <- locScoreMat[links$target, "InteractionScore"]

# 保存到文件
write.csv(links, "easy_input.csv", quote = F, row.names = F)
```

## 开始画图

用igraph画图

```{r net, fig.height=2, fig.width=2}
# 读取相互作用关系和score
links <- read.csv("easy_input.csv", header = T)
head(links)

net <- graph_from_data_frame(links, directed = F)

pal_loc <- setNames(RColorBrewer::brewer.pal(7, name = "Set3")[-2], 
                      sort(unique(links$Loc)))

# see plot.igraph for more settings 
vertexes <- get.vertex.attribute(net, "name")
net %<>% 
  set.vertex.attribute("label", 
                       value = ifelse(vertexes == sourceGene, 
                                      sourceName, vertexes)) %<>%
  set.edge.attribute("color", 
                     value = pal_loc[links$Loc]) %<>%
  set.edge.attribute("weight", 
                     value = links$score)

set.seed(777)
lay <- layout_with_dh(net) #weight.node.dist = unique(links[,c("target","score")])$score) # weight here may be invalid!

# plot in SVG device, then you may use Adobe Illustrator 
# to fine tune the network map. 
# Such as adjust the text position to avoid overlapping.

pdf(file = "interaction.pdf", width = 6, height = 6)
plot(net, layout = lay, 
     vertex.shape = 'fcircle',
     vertex.size = ifelse(vertexes == sourceGene, 15, 3),
     vertex.color = "hotpink",
     vertex.frame.color = ggplot2::alpha("white", alpha = 0.5),
     vertex.frame.width = ifelse(vertexes == sourceGene, 30, 0.1),
     vertex.label.color = ifelse(vertexes == sourceGene, "black", "grey30"),
     vertex.label.family = "sans",
     vertex.label.cex = ifelse(vertexes == sourceGene, 0.9, 0.6),
     edge.curved = curve_multiple(net, start = 0.06)) 
     # fine tune "start" to get better visual effect. To shrink multiple edges using smaller value.
legend(-1, -1, legend= names(pal_loc),
       col= pal_loc, lty=1, cex=0.6,
       box.lty=1, box.lwd=1, box.col="black")
dev.off()
```

![](interaction.pdf)

## 后期处理

Use Adobe Illustrator to fine tune the network map. Such as adjust the text position to avoid overlapping.

Want to learn more about network visualization with R?

GO to <https://kateto.net/tutorials/>

```{r}
sessionInfo()
```