---
title: "FigureYa72biomarker"
author: "小丫画图出品"
date: "2019-1-27"
output: html_document
---
欢迎关注“小丫画图”公众号，同名知识星球等你加入

小丫微信: epigenomics  E-mail: figureya@126.com

作者：王童鞋

小丫编辑校验

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 需求描述

复现原文的nearest shrunken centroid classification找biomarker，画出文章里的图。

![](example.png)

出自<https://www.nature.com/articles/s41598-017-14314-y>

## 应用场景

找癌症分型的biomarker。不限于例文的DNA甲基化数据，经典的用法是根据表达数据分型。

参考资料：Robert Tibshirani, Trevor Hastie, Balasubramanian Narasimhan, and Gilbert Chu. 2002. Diagnosis of multiple cancer types by shrunken centroids of gene expression PNAS 99: 6567-6572.

## 环境设置

使用国内镜像安装包

```r
options("repos"= c(CRAN="https://mirrors.tuna.tsinghua.edu.cn/CRAN/"))
options(BioC_mirror="http://mirrors.ustc.edu.cn/bioc/")
install.packages("pamr")
install.packages('R.utils') # required by data.table::fread to read gipped file
install.packages("data.table")
```

加载包

```{r}
library(pamr)
library(data.table)

Sys.setenv(LANGUAGE = "en") #显示英文报错信息
options(stringsAsFactors = FALSE) #禁止chr转成factor
```

## 参数设置

```{r}
cols = c("red","green")
labels = c("Solid Tissue Normal","Primary Tumor")

# 如果对电脑有信心，就都设为0，也就是不抽样，用全部数据计算
# 此处设为50，tumor和normal分别抽取50个sample
STN_number <- 50 # sampling for reduce memory, maximum usage ~5G
PT_number <- 50  # set to 0 for no sampling (huge memory usage !!!小型服务器也受不了)

set.seed(6666)
```

## 输入数据预处理

做nearest shrunken centroid classification需要两种信息：分类（例如normal和tumor）和特征数据（表达矩阵、DNA甲基化等等）。例文用TCGA KIRC的DNA甲基化芯片数据找区分tumor跟normal的biomarker。

从[UCSC xena](https://xenabrowser.net/datapages/)下载Kidney Clear Cell Carcinoma (KIRC)处理好的甲基化芯片数据和表型注释文件。下载地址：

- DNA甲基化芯片：<https://gdc.xenahubs.net/download/TCGA-KIRC/Xena_Matrices/TCGA-KIRC.methylation450.tsv.gz>
- 表型：<https://tcga.xenahubs.net/download/TCGA.KIRC.sampleMap/KIRC_clinicalMatrix.gz>

### 表型

```{r}
# 根据sample ID标注出normal和tumor
phenotype <- read.table(file = gzfile("KIRC_clinicalMatrix.gz"), header = T, sep="\t", 
                        colClasses=c("character", rep("NULL",82), "integer", rep("NULL",37)), 
                        row.names = 1)
head(phenotype)

sample <- read.table(file = gzfile("TCGA-KIRC.methylation450.tsv.gz"), sep="\t", stringsAsFactors = F, nrows=1, row.names = 1)
sample_type <- phenotype[substr(sample,1,15), "sample_type_id"]
STN_samples <- which(sample_type==11) # sample_type_id==11 means Solid Tissue Normal
PT_samples <- which(sample_type==1) # sample_type_id==1 means Primary Tumor

# 为照顾小伙伴个人电脑的内存，提供了抽样的参数设置
# 如果对电脑有信心就不抽样，全部放进去运行。
if(STN_number==0){
  STN_samples_select <- STN_samples
  STN_number <- length(STN_samples)
} else{
  STN_samples_select <- sample(STN_samples, STN_number)
}
if(PT_number==0){
  PT_samples_select <- PT_samples
  PT_samples <- length(PT_samples)
} else{
  PT_samples_select <- sample(PT_samples, PT_number)
}
```

### DNA甲基化数据

```{r}
# 读取normal和tumor各50个sample，共485577个位点
methylation_matrix <- fread(file="TCGA-KIRC.methylation450.tsv.gz", sep="\t", stringsAsFactors = F, header = T, select = c(STN_samples_select, PT_samples_select) + 1)
dim(methylation_matrix)

# 读取位点ID
cgList <- fread(file = "TCGA-KIRC.methylation450.tsv.gz", sep = "\t", stringsAsFactors = F, header = T, select = 1)
head(cgList)

# sample ID
sample_select <- substr(sample[c(STN_samples_select, PT_samples_select)], 1, 15)
head(sample_select)

# tumor和normal表型
(sample_type_select <- phenotype[sample_select, "sample_type_id"])

# 去掉在所有sample中未被检出的位点
cgList_selector <- apply(methylation_matrix, 1, function(x) {sum(is.na(x)) == 0})
cgList_select <- unlist(cgList[cgList_selector])
```

## nearest shrunken centroid classification

先把上面各种信息存到metylation.data里，然后用到pamr包里的两个函数：pamr.train和pamr.predict。

用?pamr.train查看对输入的要求

- x- an expression genes in the rows, samples in the columns)
- y- a vector of the class labels for each sample. Optional components- genenames, a vector of gene names, and geneid- a vector of gene identifiers.

用?pamr.predict查看对输入的要求

- fit. The result of a call to pamr.train
- newx. Matrix of features at which predictions are to be made
- threshold. The desired threshold value

```{r}
# pamr.train - A function that computes a nearest shrunken centroid for gene expression (microarray) data
metylation.data <- list(x = matrix(unlist(methylation_matrix[cgList_selector,]), 
                                   nrow = sum(cgList_selector), 
                                   ncol=ncol(methylation_matrix)), 
                        y = sample_type_select, 
                        genenames = cgList_select, 
                        geneid = cgList_select, 
                        samplelabels = sample_select,
                        batchlabels = NULL)
# release memory
rm(methylation_matrix,cgList); gc() 

# list train results
(metylation.train <- pamr.train(data = metylation.data))

# pamr.predict - A function giving prediction information, from a nearest shrunken centroid fit
pamr.getGenes <- function (fit, data, threshold){
  x <- data$x[fit$gene.subset, fit$sample.subset]
  geneid <- data$geneid[fit$gene.subset]
  aa <- pamr.predict(fit, x, threshold = threshold, type = "nonzero")
  g1 <- geneid[aa]
  rm(x,geneid,aa)
  gc()
  return(as.character(g1))
}
# use 17 as threshold to get top 11 cgsites
cgsites <- pamr.getGenes(metylation.train, metylation.data, threshold=17)
cgsites
```

cg25247520, cg11201447 in cgsites

## 开始画图

```{r}
for(i in 1:length(cgsites)){
  cgsite <- cgsites[i]
  methylValue <- metylation.data$x[which(cgList_select==cgsite), ]
  
  pdf(paste0("groupMethylation_",cgsite,".pdf"))
  plot(1:STN_number,methylValue[1:STN_number], pch=1, col=cols[1], main=cgsite,
       xlim=c(1,length(methylValue)), ylim=c(0,1),
       yaxs="i",xaxs="i",xlab="",xaxt="n",ylab="Methylated levels")
  points((STN_number+1):length(methylValue), methylValue[(STN_number+1):length(methylValue)], 
         pch=1, col=cols[2])
  abline(v=STN_number+0.5, lty=2, lwd=2)
  legend("topright", pch=1, col=cols, legend=labels, bty="n")
  dev.off()
}
```

在当前文件夹会看到生成了11个pdf文件，每个cgsites对应一个pdf文件。

## 原图复现

原文选出的4个位点，有两个（cg25247520和cg11201447）出现在我们找到的cgsites当中。

下面画原文选出的这4个位点。只要提供KIRC.methylation450.tsv.gz和KIRC_clinicalMatrix.gz两个文件，就能直接画图。

```{r}
cgsites = c("cg08995609", "cg25247520", "cg13309012", "cg11201447")
cols = c("red", "green")
labels = c("Normal", "Tumor")

# load phenotype & sample list
phenotype <- read.table(file=gzfile("KIRC_clinicalMatrix.gz"), 
                        header = T, sep="\t", 
                        colClasses=c("character", rep("NULL",82), "integer", rep("NULL",37)),
                        row.names = 1)

sample <- read.table(file=gzfile("TCGA-KIRC.methylation450.tsv.gz"), 
                     sep = "\t", stringsAsFactors = F, 
                     nrows = 1, #只读取第一行
                     row.names = 1)
sample_type <- phenotype[substr(sample[1,], 1, 15), "sample_type_id"]
normal_number <- sum(sample_type == 11)
isTumor <- sample_type != 11
isNormal <- sample_type == 11

# get cg list
cgList <- read.table(file = gzfile("TCGA-KIRC.methylation450.tsv.gz"),
                     sep="\t", stringsAsFactors = F, header = T,
                     colClasses = c("character", rep("NULL", ncol(sample)-1)))
str(cgList)

# 开始画图
for(i in 1:length(cgsites)){
  cgsite <- cgsites[i]
  methylValue <- read.table(file = gzfile("TCGA-KIRC.methylation450.tsv.gz"), sep = "\t",
                            skip = which(cgList[,1] %in% cgsite),
                            nrows = 1,row.names = 1)
  
  pdf(paste0("groupMethylation_", cgsite, "_4.pdf"))
  plot(1:normal_number, methylValue[1, isNormal],
       pch = 1, col = cols[1], main = cgsite,
       xlim = c(1, length(methylValue)), ylim = c(0,1),
       yaxs = "i", xaxs = "i", xlab = "", xaxt = "n", ylab = "Methylated levels")
  points((normal_number + 1):length(methylValue), methylValue[1, isTumor],
         pch=1, col=cols[2])
  abline(v=normal_number + 0.5, lty = 2, lwd = 2)
  legend("topright", pch = 1, col = cols, legend = labels,bty="n")
  dev.off()
}
```

```{r}
sessionInfo()
```