---
title: "FigureYa195PanPair"
author: "小丫画图出品"
date: "2020-9-18"
output: html_document
---
欢迎关注“小丫画图”公众号，回复“小白”，看小视频，实现点鼠标跑代码。

小丫微信: epigenomics  E-mail: figureya@126.com

作者：大鱼海棠

单位：中国药科大学国家天然药物重点实验室，生物统计与计算药学研究中心

小丫编辑校验

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述

直接比较了泛癌里面paired的基因的表达差异。之前的FigureYa55panCancer_violin比较癌和normal，没有paired的数据。

![](example.png)

出自<https://www.nature.com/articles/s41388-019-1026-9>

Fig. 1 Expression landscape of rRNA metabolism-related genes in human cancer. 
a rRNA metabolic score among all samples grouped by cancer from the TCGA. Y-axis representing rRNA metabolic score, which was calculated by ssGSEA based on the gene expression in the TCGA. Boxplots show median, quartiles, min, and max, each point representing one sample. p-values are based on the Mann–Whitneytest. 
b Similar to (a), but **in paired samples** grouped by cancer from the TCGA. Each point representing one sample. p-values are based on two-tailed Student t-test.

# 应用场景

提取TCGA泛癌里配对样本的表达量，再从中提取感兴趣的通路里的基因，计算通路富集得分，画出连线图。

根据barcode提取配对样本的方法同样适用于TCGA的其他类型数据，例如DNA甲基化等。

配对样本的表达量还可以用来做更能多分析，画更多图。例如用来筛选配对间的差异表达基因，用FigureYa149rankHeatmap来展示。

还可以把基因集换成你感兴趣的signature、biomarker。你在某个cohort里发现的差异基因，再来TCGA泛癌里看看异同。

更多泛癌的图看这里<https://k.koudai.com/Wi1xos9X>

关注“小丫画图”微信公众号，回复“195”，查看配对样本在更多文章中的展示方式。

# 环境设置

使用国内镜像安装包

```{r eval=FALSE}
options("repos"= c(CRAN="https://mirrors.tuna.tsinghua.edu.cn/CRAN/"))
options(BioC_mirror="http://mirrors.tuna.tsinghua.edu.cn/bioconductor/")
install.packages(c("data.table", "knitr"))
```

加载包

```{r}
library(data.table)
library(GSVA)
Sys.setenv(LANGUAGE = "en") #显示英文报错信息
options(stringsAsFactors = FALSE) #禁止chr转成factor
```

# 输入文件的获得

如果你已经准备好easy_input_pair.txt、easy_input_expr.RDS和easy_input_gene.txt文件，就可以跳过这步，直接进入“输入文件”。

## TCGA pan-cancer配对样本的表达矩阵提取

这步运行一次就行，可直接跳到“获取感兴趣的通路里的基因名”。

这步提取出的配对样本及其表达矩阵，已保存到easy_input_pair.txt和easy_input_expr.RDS文件。你可以用这些配对样本的表达量做很多事情，不要局限于例图哦～

GDC-PANCAN.basic_phenotype.tsv，样本信息，从XENA下载：<https://xenabrowser.net/datapages/?dataset=GDC-PANCAN.basic_phenotype.tsv&host=https%3A%2F%2Fgdc.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443>

EBPlusPlusAdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp.tsv，表达矩阵，从TCGA PanCanAtlas下载：<https://gdc.cancer.gov/about-data/publications/pancanatlas>。太大，已上传微云<https://share.weiyun.com/K5dlEurP>

```{r eval=FALSE}
# 载入样本信息
pheno <- read.table("GDC-PANCAN.basic_phenotype.tsv",sep = "\t",row.names = 1,header = T,check.names = F,stringsAsFactors = F)

# 加载基因表达矩阵
expr <- fread("EBPlusPlusAdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp.tsv",sep = "\t",header = T,check.names = F,stringsAsFactors = F)

# 表达矩阵数据处理
expr <- as.data.frame(expr)
expr$gene_id <- sapply(strsplit(expr$gene_id,"|",fixed = T), "[", 1) # 取|左侧的基因名
expr <- expr[!duplicated(expr$gene_id),] # 去重
rownames(expr) <- expr$gene_id; expr <- expr[-grep("?",rownames(expr),fixed = T),-1] # 移除带有?的基因
colnames(expr) <- substr(colnames(expr), start = 1, stop = 16) # 取前16位barcode
expr[1:3,1:3]

# 样本信息处理
pheno <- pheno[which(pheno$program == "TCGA" & pheno$sample_type_id %in% c(1,11)),] # 取TCGA中原位癌和正常样本

# 更新数据
com_sam <- intersect(colnames(expr),rownames(pheno))
expr <- expr[,com_sam]
pheno <- pheno[com_sam,]
```

根据TCGA sample barcode的规则提取配对样本及其对应的表达矩阵，<https://docs.gdc.cancer.gov/Encyclopedia/pages/TCGA_Barcode/>

```{r eval=FALSE}
proid <- unique(pheno$project_id)
tcga.pan.pair <- NULL
for (i in proid) {
  sam <- rownames(pheno[which(pheno$project_id == i),])
  n.sam <- sam[grep("-11A",sam)]
  t.sam <- sam[grep("-01A",sam)]
  
  if(length(n.sam) > 0) {
    toMatch <- substr(n.sam,start = 9, stop = 12)
    matches <- unique(grep(paste(toMatch,collapse = "|"), 
                           t.sam, value = TRUE))
    
    if(length(matches) > 0) {
      # check if they are identical
      t.code <- substr(matches, start = 1, stop = 12)
      n.code <- substr(n.sam, start = 1, stop = 12)
      com_code <- intersect(t.code, n.code)
      
      pair.t.sam <- paste0(com_code, "-01A")
      pair.n.sam <- paste0(com_code, "-11A")
      
      tmp <- data.frame(samID = c(pair.t.sam,pair.n.sam),
                        tissue = rep(c("paired.tumor","paired.normal"), each = length(com_code)),
                        project = i,
                        row.names = c(pair.t.sam,pair.n.sam),
                        stringsAsFactors = F)
      tcga.pan.pair <- rbind.data.frame(tcga.pan.pair,tmp,stringsAsFactors = F)
    } else {cat("Fail to find matched tumor samples for ",i,"\n")}

  } else {cat("Fail to find normal samples for ",i,"\n")}
}
table(tcga.pan.pair$project)

# 把配对样本信息及其表达矩阵保存到文件
write.table(tcga.pan.pair,"easy_input_pair.txt",sep = "\t",row.names = F,quote = F)

# 进一步更新数据
expr <- as.data.frame(na.omit(expr[,tcga.pan.pair$samID]))
# 把配对样本的表达矩阵保存到文件
saveRDS(expr, "easy_input_expr.RDS")
```

## 获取感兴趣的基因集

感兴趣的通路里包含哪些基因？

可以从领域内的网站或数据库下载，或者通过阅读综述、研究论文，自己总结基因名。

这里提供一个方法，从GSEA网站下载MSigDB的gmt文件<http://software.broadinstitute.org/gsea/downloads.jsp>，从gmt文件中获得感兴趣通路里的基因，以`h: hallmark gene sets`为例，下载`h.all.v7.1.symbols.gmt`文件。

用到了FigureYa151pathifier里的自定义函数

```{r eval=FALSE}
# 自定义函数，用于读取gmt文件为列表形式
gmt2list <- function(annofile){
  if (!file.exists(annofile)) {
    stop("There is no such gmt file.")
  }
  
  if (tools::file_ext(annofile) == "xz") {
    annofile <- xzfile(annofile)
    x <- scan(annofile, what="", sep="\n", quiet=TRUE)
    close(annofile)
  } else if (tools::file_ext(annofile) == "gmt") {
    x <- scan(annofile, what="", sep="\n", quiet=TRUE)
  } else {
    stop ("Only gmt and gmt.xz are accepted for gmt2list")
  }
  
  y <- strsplit(x, "\t")
  names(y) <- sapply(y, `[[`, 1)
  
  annoList <- lapply(y, `[`, c(-1,-2))
}

# 读取gmt文件
gset <- gmt2list("h.all.v7.1.symbols.gmt") 

# 提取带有OXIDATIVE字样的通路
gset_meta <- gset[names(gset) %like% "OXIDATIVE"]
gset_meta
meta <- gset_meta$HALLMARK_OXIDATIVE_PHOSPHORYLATION

# 或者要多个通路
#meta <- c(gset$HALLMARK_DNA_REPAIR, gset$HALLMARK_APOPTOSIS)

# 保存到文件
#write.table(meta, "easy_input_gene.txt", row.names = F, col.names = F, quote = F)
```

# 输入文件

easy_input_gene.txt，代谢基因，来自例文补充材料41388_2019_1026_MOESM2_ESM.xlsx中Table S1。不局限于通路，还可以是某一群基因，例如前期研究筛选出的signature等。

easy_input_pair.txt，配对样本信息。

easy_input_expr.RDS，配对样本的表达矩阵。太大，已上传微云<https://share.weiyun.com/K5dlEurP>

```{r}
# 加载感兴趣通路里的基因
meta <- read.table("easy_input_gene.txt")$V1

# 加载配对样本信息
tcga.pan.pair <- read.table("easy_input_pair.txt", header = T)

# 加载配对样本的表达矩阵
expr <- readRDS("easy_input_expr.RDS")
expr[1:3, 1:3]
```

# 单样本富集得分

大白话：把每一个样本中多个基因的表达值（多个数值）用ssGSEA换算成一个数值。这样一来，画图时就可以清晰地看到每对配对样本的normal和tumor之间的表达量谁高谁低了。

用ssGSEA计算通路的富集得分

```{r}
indata <- expr
indata <- round(log2(pmax(indata + abs(min(indata)), 0) + .Machine$double.eps + 1),3) # 每个数加上最小值的绝对值后取对数

# 运行ssgsea
meta.ssgsea <- gsva(expr = as.matrix(indata),
                    gset.idx.list = list("metabolism" = meta),
                    method = "ssgsea",
                    parallel.sz = 1)
pan.meta <- tcga.pan.pair
pan.meta$mscore <- scale(as.numeric(meta.ssgsea[1,pan.meta$samID])) # z-score变换
```

# 配对检验

```{r}
proid <- unique(pan.meta$project)
wp <- sig <- c()
for(i in proid) {
  
  tmp <- pan.meta[which(pan.meta$project == i),]
  t <- tmp[which(tmp$tissue == "paired.tumor"),]
  n <- tmp[which(tmp$tissue == "paired.normal"),]
  wp <- c(wp, wilcox.test(t$mscore,n$mscore,paired = T)$p.value)
  sig <- c(sig, as.character(cut(wilcox.test(t$mscore,n$mscore,paired = T)$p.value,
                                 c(0,0.001,0.005,0.01,0.05,1),c("****","***","**","*",""))))
}
names(wp) <- names(sig) <- unique(pan.meta$project)
wp
sig
```

# 开始画图

```{r}
# baseplot绘图 #
ylim <- c(floor(range(pan.meta$mscore)[1]), ceiling(range(pan.meta$mscore)[2]))

pdf("Metabolic landscape of tcga paired sample.pdf",width = 8.5,height = 3.5)
showLayout <- F # 默认不在最终pdf的首页显示layout结构，不过建议初次绘制的时候改为TRUE看一下，方便理解

# 设置画面布局，相同数字代表同一区块，数字越多代表该区块所占面积越大
mat <- c(rep(c(rep(1,2), rep(2,2), rep(3:24,each = 3)),16))
layout(matrix(c(mat, rep(25, 4+22*3)),byrow = T,nrow = 17)) # 请输出这个matrix理解一下
# print(matrix(c(mat, rep(25, 4+22*3)),byrow = T,nrow = 17))

if(showLayout) {
  layout.show(n = 25) # 直观展示画布分布
}

#---------------------------#
# 画布区域1：绘制最左侧标签 #
#---------------------------#
par(bty = "n", mgp = c(2,.6,0), mar = c(0,0,0,0), xpd = T)
plot(0,0,col = "white",
     xlab = "",xaxt = "n", # 不显示x坐标轴
     ylab = "",yaxt = "n",
     xaxs = "i", yaxs = "i") # 不显示y坐标轴
text(0,0,"rRNA metabolic score\n[Z-score (ssGSE)]", srt = 90, cex = 1, font = 2)

#------------------------------#
# 画布区域2：绘制最左侧区域y轴 #
#------------------------------#

# 因为区域太小所以无法和1合并，所以先画一个y轴标签，再画一个y轴
par(bty = "o", mgp = c(2,.6,0), mar = c(3,1.5,1,0), las = 2, font.axis = 2, xpd = F) # 基础参数
a <- barplot(c(0,0),
             border = NA,
             ylim = ylim,
             xlab = "", xaxt = "n", # 不显示x坐标轴
             ylab = "rRNA metabolic score\n[Z-score (ssGSE)]", yaxt = "n") # 不显示y坐标轴
axis(side = 2, at = pretty(ylim), lwd = 2)
abline(h = par("usr")[3], col = "black", lwd = 3)
abline(h = par("usr")[4], col = "black", lwd = 2) 

#----------------------------#
# 画布区域3-24：绘制22种癌症 #
#----------------------------#
for (i in proid) {
  
  # 取出肿瘤和正常的代谢评分
  t.dat <- pan.meta[which(pan.meta$project == i & pan.meta$tissue == "paired.tumor"),]
  n.dat <- pan.meta[which(pan.meta$project == i & pan.meta$tissue == "paired.normal"),]
  
  # 先画一个无色的柱状图确定x轴上两个对齐的位置，保存在a对象里
  par(bty = "n", mgp = c(2,.6,0), mar = c(3,0,1,0), las = 2, font.axis = 2, xpd = F) # 基础参数
  a <- barplot(c(0,0), col = "white",
               border = NA,
               ylim = ylim,
               xlab = "", xaxt = "n", # 不显示x坐标轴
               ylab = "", yaxt = "n") # 不显示y坐标轴
  
  # 在x轴对齐位置的中心点，且画布的最上方标记显著性（参考par("usr")的用法）
  par(xpd = T)
  text((a[1,] + a[2,])/2 - 0.05, y = par("usr")[3]-0.4, gsub("TCGA-","",i), srt = 35) # 倾斜35度角
  abline(h = par("usr")[3], col = "black", lwd = 2)
  abline(h = par("usr")[4], col = "black", lwd = 2) 
  par(xpd = F)
  
  # 如果此时是最后一个癌症，则在小区块右侧封上黑色实线
  if(i != proid[length(proid)]) {
    abline(v = par("usr")[2], col = "#3B50A0", lwd = 3, lty = 3) 
  } else { # 否则封上蓝色虚线
    abline(v = par("usr")[2]-0.05, col = "black", lwd = 3) 
  }
  
  # 先画出配对样本的蓝色连线，如果后画，连线的端点会盖住下面的圆圈
  segments(a[1,], t.dat$mscore,
           a[2,], n.dat$mscore,
           col = "#6C99CE", lwd = 1.5)
  
  # 再画圆圈，以遮盖住连线的端点
  points(rep(a[1,], nrow(t.dat)), t.dat$mscore, pch = 19, col = "#BB61A1", cex = 1.2)
  points(rep(a[2,], nrow(n.dat)), n.dat$mscore, pch = 19, col = "#7EC09A", cex = 1.2)
  
  # 在x轴对齐位置的中心点下方位置，标记癌症名称
  text((a[1,] + a[2,])/2, y = par("usr")[4]-0.3, as.character(sig[i]), cex = 1.2)
}

#--------------------------#
# 画布区域25：绘制底部图例 #
#--------------------------#

par(bty = "n", mgp = c(2,.6,0), mar = c(0,0,0,0), xpd = T)
plot(0,0,col = "white",
     xlab = "",xaxt = "n", # 不显示x坐标轴
     ylab = "",yaxt = "n",
     xaxs = "i", yaxs = "i") # 不显示y坐标轴
legend((par("usr")[1] + par("usr")[2])/2,par("usr")[4],
       legend = c("Tumor","Normal"),
       col = c("#BB61A1","#7EC09A"),
       pch = c(19,19),
       border = NA, # 图例颜色没有边框
       bty = "n", # 图例没有边框
       cex = 1.1,
       x.intersp = 0.5,
       y.intersp = 1,
       #yjust = 0.5,
       horiz = T) # 图例水平放置

# 关闭图形句柄
invisible(dev.off())

# 保存对象
# save.image("PanMeta.RData")
```

![](Metabolic landscape of tcga paired sample.pdf)

# Session Info

```{r}
sessionInfo()
```