---
title: "FigureYa44 profile"
author: "小丫画图"
date: "2018-10-8"
output: html_document
---
微信ID: epigenomics  E-mail: figureya@126.com

作者：王童鞋

小丫编辑校验

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 需求描述

用R画出两个样品对比的profile。

![](example.png)


Figure 3 | Genome-wide maps of Cxxc1 binding and H3K4me3 modification in DP thymocytes.

(b) Distribution of Cxxc1-binding peaks across extended gene bodies. The tag density of Cxxc1 binding was calculated on gene bodies (between transcription start site, TSS and transcription termination site, TTS), as well as 3-kb upstream of TSS and 3-kb downstream of TTS regions of all RefSeq (mm9) genes. Uniquely, mapped tags were summarized in 100-bp windows for promoter regions and in 5% of gene body sequences. All window tag counts were normalized by the total number of bases in the windows and the total read number of the given sample.

(c) Enrichment of Cxxc1-binding peaks on CpG islands (CpGI). The tag density of Cxxc1 binding was calculated on CpG islands and 5-kb flanking regions.

出自<https://www.nature.com/articles/ncomms11687>

## 应用场景

对比不同的样品在某一区域的信号特征，不仅限于ChIP-seq、DNase-seq、ATAC-seq数据。

对比IP实验和input对照的信号强度，例如需求描述里的图：

(b) 把基因压缩到统一长度，观察转录起始位点（TSS）、转录终止位点（TTS）、gene body以及两侧的信号特征。

(c) 观察某一功能区域，例如CpGi、TSS、TTS、peak summits或enhancer区及其两侧的信号特征。

## b图

为了提取bw文件里的信号，需要用到2个函数`regionCapture`、`signal_caputer_around_gene`，保存在`corelib.R`文件里

```{r}
#install.packages("rtracklayer")
source("./corelib.R") #位于当前文件夹
```

### 参数设置

在这里修改常见的需要调整的参数

```{r}
outputName <- "geneBodyProfiler_example.pdf" #输出的pdf文件名
bigWigFiles <- c("cxxc1.bw","control.bw") #bw文件，位于当前文件夹，可以>2个
labels <- c("Cxxc1","Input") #图例标签，跟bw数量一致
colors <- c("navy","firebrick") #线的颜色，跟bw数量一致

normalization_constant <- 1 #均一化的系数，根据展示所需信号值刻度调整
genome <- "mm9" #跟bw文件一致的基因组版本
resolution <- 100 #分辨率，数值越低，画出来的曲线越平滑，运行时间越长
span <- 3000 #上下游展示范围，需要是resolution的整数倍
geneBodySpan <- 2000 #gene body normalization length，需要是resolution的整数倍
coreNumber <- detectCores() - 1 #留出一个核用来干别的事情
```

### 输入文件

cxxc1.bw和control.bw：ChIP-seq样品及其contorl的bw文件。在生成bw时做过normalization。

easy_input.txt：想要查看的基因的位置信息，至少包含4列：染色体、正负链、起始位点、终止位点。作者用的是refGene里所有经过验证的编码RNA（NM开头），此处考虑到运行时间，从中随机抽取5000个基因。

如果你要提取的基因已经保存成类似easy_input.txt的格式，就可以跳过这步，直接进入“信号提取”。

```{r}
#提前下载好refGene.gz文件
#download.file(paste0("http://hgdownload.soe.ucsc.edu/goldenPath/",genome,"/database/refGene.txt.gz"),"refGene.gz")
refGene <- read.table(gzfile("refGene.gz"))

isNormalChrosome <- !grepl("_",refGene$V3)    # remove abnormal chromsome
ismRNA <- grepl("^NM",refGene$V2)    # only keep valid mRNAs
isEnoughtLength <- (refGene[,6] - refGene[,5]) > geneBodySpan / resolution

flag <- isNormalChrosome & ismRNA & isEnoughtLength
refGene <- refGene[flag,]

# sampling, remove these part for final figure
set.seed(6666)
idx <- sample(1:nrow(refGene), size = 5000)
myGene <- refGene[idx,]
str(myGene[1:6])
write.table(myGene[1:6], "easy_input.txt", quote = F, sep = "\t", row.names = F, col.names = F)
#delete_result <- file.remove("refGene.gz")
```

### 信号提取

提取的基因越多，运行的时间越长。

此处提取了5000个基因的信号，需要十分钟左右

```r
myGene <- read.table("easy_input.txt", as.is = T)
head(myGene)
str(myGene)
average_signal <- c()

for(i in 1:length(bigWigFiles)){
  average_signal <- rbind(average_signal,
                          signal_caputer_around_gene(bigWigFiles[i], #要提取的bw文件
                                   myGene, #如果提取全部基因的信号，就换成refGene
                                   resolution = resolution,
                                   span = span,
                                   geneBodySpan = geneBodySpan,
                                   cores = as.integer(coreNumber)))
}
average_signal <- average_signal / normalization_constant #均一化的系数，根据测序深度调整

#这步比较耗时，我们把抽取的信号保存到`gene_average_signal.csv`文件里。
write.table(average_signal,"gene_average_signal.txt",quote = F,row.names = F,col.names = F)
```

### 开始画图

```{r, fig.width=6, fig.height=4}
average_signal <- read.table("gene_average_signal.txt", header = F, as.is = T)
minimum <- min(average_signal) - (max(average_signal) - min(average_signal)) * 0.1
maximum <- max(average_signal) + (max(average_signal) - min(average_signal)) * 0.2

datapoint1 <- span / resolution
datapoint2 <- geneBodySpan / resolution
total_length <- datapoint1 * 2 + datapoint2

#pdf(outputName,width = 6,height = 4)
plot(1:total_length, average_signal[1,], 
     type = "l", lwd = 3, tck = 0, #不显示刻度线
     mgp = c(2,0.5,0), #坐标轴标题、刻度和坐标轴线跟作图区域的距离
     col = colors[1],
     xaxt = "n", #不显示x轴刻度
     xlab = "", ylab = "Normalized signal",
     ylim = c(minimum ,maximum))
for(i in 2:length(bigWigFiles)){
    lines(1:total_length, average_signal[i,], col = colors[i], lwd = 3) #线的颜色、粗细
}

#在TSS和TTS的位置画两条虚线
#abline(v = c(datapoint1, datapoint1 + datapoint2) + 0.5, lty = 2)

legend("topright", col = colors, ncol = 2,
       legend = labels, lty = 1, lwd = 3,
       bty = "n")#不显示图例边框

box(lwd = 2) #边框粗细

axis(side = 1, tck = 0, #不显示刻度线
     mgp = c(2,0.5,0),
     at = c(0, datapoint1, datapoint1 + datapoint2/2, datapoint1 + datapoint2, total_length) + 0.5,
     labels = c(paste(-1*span/1000, "kb"), "TSS", "50%", "TTS", paste(span/1000, "kb")))
     
#dev.off()
```

## c图

同样需要`corelib.R`文件提取bw文件里的信号，用到2个函数：`regionCapture`、`signal_caputer_around_sites`

```{r}
source("./corelib.R") #位于当前文件夹
```

### 参数设置

在这里修改常见的需要调整的参数

```{r}
outputName <- "sitepro_example.pdf"
bigWigFiles <- c("cxxc1.bw","control.bw") 
labels <- c("Cxxc1","Input")
colors <- c("navy","firebrick") 

bedFile <- "zv9_CpG_islands.bed" #要画的功能区域的bed文件
siteType <- "CpG island"
resolution <- 100
span <- 5000
normalization_constant <- 1
coreNumber <- detectCores() - 1
```

### 输入文件

cxxc1.bw和control.bw：ChIP-seq样品及其contorl的bw文件，在生成bw时做过normalization。

zv9_CpG_islands.bed：想要查看的基因/功能区域的bed文件，此处是15000个CpGi。

### 提取信号

15000个CpGi，运行时间是上一个图的2-3倍。

```r
average_signal <- c()
for(i in 1:length(bigWigFiles)){
  average_signal <- rbind(average_signal,
                          signal_caputer_around_sites(bigWigFiles[i], #要提取信号的bw文件
                                                      bedFile, #要提取的区域列表
                                                      resolution=resolution,
                                                      span=span,
                                                      cores=as.integer(coreNumber)))
}
average_signal <- average_signal / normalization_constant
#这步比较耗时，我们把抽取的信号保存到`site_average_signal.csv`文件里。
write.table(average_signal,"site_average_signal.txt",quote = F,row.names = F,col.names = F)
```

### 开始画图

```{r, fig.width=6, fig.height=4}
average_signal <- read.table("site_average_signal.txt", header = F, as.is = T)
minimum <- min(average_signal) - (max(average_signal) - min(average_signal)) * 0.1
maximum <- max(average_signal) + (max(average_signal) - min(average_signal)) * 0.2

datapoints <- span / resolution

#pdf(outputName,width = 6,height = 4)
plot(1:(datapoints*2+1), average_signal[1,], 
     type="l", lwd = 3, tck = 0, #不显示刻度线
     mgp = c(2,0.5,0), #坐标轴标题、刻度和坐标轴线跟作图区域的距离
     col=colors[1],
     xaxt = "n", #不显示x轴刻度
     xlab="",ylab="Normalized signal",ylim=c(minimum,maximum))
for(i in 2:length(bigWigFiles)){
    lines(1:(datapoints*2+1), average_signal[i,], col = colors[i], lwd = 3) #线的颜色、粗细
}

legend("topleft", col = colors, legend = labels, lty = 1, lwd = 3,
       bty = "n")#不显示图例边框

box(lwd = 2) #边框粗细

axis(side = 1, tck = 0,
     mgp = c(2,0.5,0),
     at = c(0, datapoints, datapoints*2) + 1,
     labels = c(paste(-1*span/1000, "kb"), siteType, paste(span/1000, "kb")))
#dev.off()
```

```{r}
sessionInfo()
```